{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metadata Vignette\n",
    "\n",
    "## Counting HCA attribute usage\n",
    "\n",
    "### By Matthew Green\n",
    "\n",
    "\n",
    "The HCA has tightly controlled, well defined, curated metadata assiciated with all it's projects. There are nearly 600 fields defined by our JSON metadata. But which ones are most frequently used? This vignette answers this question by doing the following:\n",
    "\n",
    "This code does the following:\n",
    "\n",
    "1. Fetches latest attributes from metadata schema\n",
    "1. Defines an example of an ES query for specific bundles (in the provided example we return all projects that are not 'black listed') and makes a generator from the DSS for bundle iteration\n",
    "1. Counts these attributes in the bundles from the geenrator\n",
    "1. Makes a few graphs and tables (there are options to save these."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.\n",
    "\n",
    "This section queries the metadata API to get the fields from the latest version of the schema. The list returned are the programatic names of the attributes. \n",
    "\n",
    "**The Query against the metadata API may take a couple of minutes.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "590 out of 590 attributes are unique\n",
      "['sequencing_protocol.protocol_core.protocol_id', 'sequencing_protocol.protocol_core.protocol_name', 'sequencing_protocol.protocol_core.protocol_description', 'sequencing_protocol.protocol_core.publication_doi', 'sequencing_protocol.protocol_core.protocols_io_doi', 'sequencing_protocol.protocol_core.document', 'sequencing_protocol.instrument_manufacturer_model.text', 'sequencing_protocol.instrument_manufacturer_model.ontology', 'sequencing_protocol.instrument_manufacturer_model.ontology_label', 'sequencing_protocol.local_machine_name', 'sequencing_protocol.paired_end', 'sequencing_protocol.method.text', 'sequencing_protocol.method.ontology', 'sequencing_protocol.method.ontology_label', 'sequencing_protocol.10x.fastq_method', 'sequencing_protocol.10x.fastq_method_version', 'sequencing_protocol.10x.pooled_channels', 'sequencing_protocol.10x.drop_uniformity', 'library_preparation_protocol.protocol_core.protocol_id', 'library_preparation_protocol.protocol_core.protocol_name', 'library_preparation_protocol.protocol_core.protocol_description', 'library_preparation_protocol.protocol_core.publication_doi', 'library_preparation_protocol.protocol_core.protocols_io_doi', 'library_preparation_protocol.protocol_core.document', 'library_preparation_protocol.cell_barcode.barcode_read', 'library_preparation_protocol.cell_barcode.barcode_offset', 'library_preparation_protocol.cell_barcode.barcode_length', 'library_preparation_protocol.cell_barcode.white_list_file', 'library_preparation_protocol.input_nucleic_acid_molecule.text', 'library_preparation_protocol.input_nucleic_acid_molecule.ontology', 'library_preparation_protocol.input_nucleic_acid_molecule.ontology_label', 'library_preparation_protocol.nucleic_acid_source', 'library_preparation_protocol.library_construction_method.text', 'library_preparation_protocol.library_construction_method.ontology', 'library_preparation_protocol.library_construction_method.ontology_label', 'library_preparation_protocol.library_construction_kit.retail_name', 'library_preparation_protocol.library_construction_kit.catalog_number', 'library_preparation_protocol.library_construction_kit.manufacturer', 'library_preparation_protocol.library_construction_kit.lot_number', 'library_preparation_protocol.library_construction_kit.expiry_date', 'library_preparation_protocol.library_construction_kit.kit_titer', 'library_preparation_protocol.nucleic_acid_conversion_kit.retail_name', 'library_preparation_protocol.nucleic_acid_conversion_kit.catalog_number', 'library_preparation_protocol.nucleic_acid_conversion_kit.manufacturer', 'library_preparation_protocol.nucleic_acid_conversion_kit.lot_number', 'library_preparation_protocol.nucleic_acid_conversion_kit.expiry_date', 'library_preparation_protocol.nucleic_acid_conversion_kit.kit_titer', 'library_preparation_protocol.end_bias', 'library_preparation_protocol.primer', 'library_preparation_protocol.strand', 'library_preparation_protocol.spike_in_kit.retail_name', 'library_preparation_protocol.spike_in_kit.catalog_number', 'library_preparation_protocol.spike_in_kit.manufacturer', 'library_preparation_protocol.spike_in_kit.lot_number', 'library_preparation_protocol.spike_in_kit.expiry_date', 'library_preparation_protocol.spike_in_kit.kit_titer', 'library_preparation_protocol.spike_in_dilution', 'library_preparation_protocol.umi_barcode.barcode_read', 'library_preparation_protocol.umi_barcode.barcode_offset', 'library_preparation_protocol.umi_barcode.barcode_length', 'library_preparation_protocol.umi_barcode.white_list_file', 'library_preparation_protocol.library_preamplification_method.text', 'library_preparation_protocol.library_preamplification_method.ontology', 'library_preparation_protocol.library_preamplification_method.ontology_label', 'library_preparation_protocol.cdna_library_amplification_method.text', 'library_preparation_protocol.cdna_library_amplification_method.ontology', 'library_preparation_protocol.cdna_library_amplification_method.ontology_label', 'library_preparation_protocol.nominal_length', 'library_preparation_protocol.nominal_sdev', 'imaging_protocol.protocol_core.protocol_id', 'imaging_protocol.protocol_core.protocol_name', 'imaging_protocol.protocol_core.protocol_description', 'imaging_protocol.protocol_core.publication_doi', 'imaging_protocol.protocol_core.protocols_io_doi', 'imaging_protocol.protocol_core.document', 'imaging_protocol.microscope_setup_description', 'imaging_protocol.microscopy_technique.text', 'imaging_protocol.microscopy_technique.ontology', 'imaging_protocol.microscopy_technique.ontology_label', 'imaging_protocol.magnification', 'imaging_protocol.numerical_aperture', 'imaging_protocol.immersion_medium_type', 'imaging_protocol.immersion_medium_refractive_index', 'imaging_protocol.pixel_size', 'imaging_protocol.number_of_tiles', 'imaging_protocol.tile_size_y', 'imaging_protocol.tile_size_x', 'imaging_protocol.z_stack_step_size', 'imaging_protocol.number_of_z_steps', 'imaging_protocol.overlapping_tiles', 'imaging_protocol.channel.channel_id', 'imaging_protocol.channel.excitation_wavelength', 'imaging_protocol.channel.filter_range', 'imaging_protocol.channel.multiplexed', 'imaging_protocol.channel.target_fluorophore', 'imaging_protocol.channel.exposure_time', 'imaging_protocol.target.molecule_name', 'imaging_protocol.target.molecule_id', 'imaging_protocol.target.subcellular_structure.text', 'imaging_protocol.target.subcellular_structure.ontology', 'imaging_protocol.target.subcellular_structure.ontology_label', 'imaging_protocol.target.reagent_name', 'imaging_protocol.target.purchased_reagent_details.retail_name', 'imaging_protocol.target.purchased_reagent_details.catalog_number', 'imaging_protocol.target.purchased_reagent_details.manufacturer', 'imaging_protocol.target.purchased_reagent_details.lot_number', 'imaging_protocol.target.purchased_reagent_details.expiry_date', 'imaging_protocol.target.purchased_reagent_details.kit_titer', 'imaging_protocol.target.probe_sequence', 'imaging_protocol.target.fluorophore', 'imaging_protocol.target.assay_type.text', 'imaging_protocol.target.assay_type.ontology', 'imaging_protocol.target.assay_type.ontology_label', 'imaging_protocol.target.multiplexed', 'imaging_protocol.target.channel_id', 'imaging_preparation_protocol.protocol_core.protocol_id', 'imaging_preparation_protocol.protocol_core.protocol_name', 'imaging_preparation_protocol.protocol_core.protocol_description', 'imaging_preparation_protocol.protocol_core.publication_doi', 'imaging_preparation_protocol.protocol_core.protocols_io_doi', 'imaging_preparation_protocol.protocol_core.document', 'imaging_preparation_protocol.fresh_slicing_method', 'imaging_preparation_protocol.imaged_slice_thickness', 'imaging_preparation_protocol.final_slicing_method', 'imaging_preparation_protocol.post_resection_interval', 'imaging_preparation_protocol.post_resection_interval_unit.text', 'imaging_preparation_protocol.post_resection_interval_unit.ontology', 'imaging_preparation_protocol.post_resection_interval_unit.ontology_label', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_method', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time_unit.text', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time_unit.ontology', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time_unit.ontology_label', 'imaging_preparation_protocol.pre_final_slice_preservation_method.preservation_method', 'imaging_preparation_protocol.post_final_slicing_interval', 'imaging_preparation_protocol.post_final_slicing_interval_unit.text', 'imaging_preparation_protocol.post_final_slicing_interval_unit.ontology', 'imaging_preparation_protocol.post_final_slicing_interval_unit.ontology_label', 'imaging_preparation_protocol.fiducial_marker', 'imaging_preparation_protocol.expansion_factor', 'ipsc_induction_protocol.protocol_core.protocol_id', 'ipsc_induction_protocol.protocol_core.protocol_name', 'ipsc_induction_protocol.protocol_core.protocol_description', 'ipsc_induction_protocol.protocol_core.publication_doi', 'ipsc_induction_protocol.protocol_core.protocols_io_doi', 'ipsc_induction_protocol.protocol_core.document', 'ipsc_induction_protocol.ipsc_induction_method', 'ipsc_induction_protocol.ipsc_induction_factors', 'ipsc_induction_protocol.ipsc_induction_kit.retail_name', 'ipsc_induction_protocol.ipsc_induction_kit.catalog_number', 'ipsc_induction_protocol.ipsc_induction_kit.manufacturer', 'ipsc_induction_protocol.ipsc_induction_kit.lot_number', 'ipsc_induction_protocol.ipsc_induction_kit.expiry_date', 'ipsc_induction_protocol.ipsc_induction_kit.kit_titer', 'ipsc_induction_protocol.pluripotency_test', 'ipsc_induction_protocol.percent_pluripotency', 'ipsc_induction_protocol.pluripotency_vector_removed', 'ipsc_induction_protocol.ipsc_induction_produced_in_house', 'ipsc_induction_protocol.protocol_reagents.retail_name', 'ipsc_induction_protocol.protocol_reagents.catalog_number', 'ipsc_induction_protocol.protocol_reagents.manufacturer', 'ipsc_induction_protocol.protocol_reagents.lot_number', 'ipsc_induction_protocol.protocol_reagents.expiry_date', 'ipsc_induction_protocol.protocol_reagents.kit_titer', 'enrichment_protocol.protocol_core.protocol_id', 'enrichment_protocol.protocol_core.protocol_name', 'enrichment_protocol.protocol_core.protocol_description', 'enrichment_protocol.protocol_core.publication_doi', 'enrichment_protocol.protocol_core.protocols_io_doi', 'enrichment_protocol.protocol_core.document', 'enrichment_protocol.enrichment_method.text', 'enrichment_protocol.enrichment_method.ontology', 'enrichment_protocol.enrichment_method.ontology_label', 'enrichment_protocol.markers', 'enrichment_protocol.min_size_selected', 'enrichment_protocol.max_size_selected', 'dissociation_protocol.protocol_core.protocol_id', 'dissociation_protocol.protocol_core.protocol_name', 'dissociation_protocol.protocol_core.protocol_description', 'dissociation_protocol.protocol_core.publication_doi', 'dissociation_protocol.protocol_core.protocols_io_doi', 'dissociation_protocol.protocol_core.document', 'dissociation_protocol.method.text', 'dissociation_protocol.method.ontology', 'dissociation_protocol.method.ontology_label', 'dissociation_protocol.reagents.retail_name', 'dissociation_protocol.reagents.catalog_number', 'dissociation_protocol.reagents.manufacturer', 'dissociation_protocol.reagents.lot_number', 'dissociation_protocol.reagents.expiry_date', 'dissociation_protocol.reagents.kit_titer', 'differentiation_protocol.protocol_core.protocol_id', 'differentiation_protocol.protocol_core.protocol_name', 'differentiation_protocol.protocol_core.protocol_description', 'differentiation_protocol.protocol_core.publication_doi', 'differentiation_protocol.protocol_core.protocols_io_doi', 'differentiation_protocol.protocol_core.document', 'differentiation_protocol.differentiation_method', 'differentiation_protocol.media', 'differentiation_protocol.small_molecules', 'differentiation_protocol.target_cell_yield', 'differentiation_protocol.reagents.retail_name', 'differentiation_protocol.reagents.catalog_number', 'differentiation_protocol.reagents.manufacturer', 'differentiation_protocol.reagents.lot_number', 'differentiation_protocol.reagents.expiry_date', 'differentiation_protocol.reagents.kit_titer', 'differentiation_protocol.target_pathway', 'differentiation_protocol.validation_method', 'differentiation_protocol.validation_result', 'collection_protocol.protocol_core.protocol_id', 'collection_protocol.protocol_core.protocol_name', 'collection_protocol.protocol_core.protocol_description', 'collection_protocol.protocol_core.publication_doi', 'collection_protocol.protocol_core.protocols_io_doi', 'collection_protocol.protocol_core.document', 'collection_protocol.method.text', 'collection_protocol.method.ontology', 'collection_protocol.method.ontology_label', 'collection_protocol.reagents.retail_name', 'collection_protocol.reagents.catalog_number', 'collection_protocol.reagents.manufacturer', 'collection_protocol.reagents.lot_number', 'collection_protocol.reagents.expiry_date', 'collection_protocol.reagents.kit_titer', 'aggregate_generation_protocol.protocol_core.protocol_id', 'aggregate_generation_protocol.protocol_core.protocol_name', 'aggregate_generation_protocol.protocol_core.protocol_description', 'aggregate_generation_protocol.protocol_core.publication_doi', 'aggregate_generation_protocol.protocol_core.protocols_io_doi', 'aggregate_generation_protocol.protocol_core.document', 'aggregate_generation_protocol.formation_method', 'aggregate_generation_protocol.cell_uniformity', 'analysis_protocol.protocol_core.protocol_id', 'analysis_protocol.protocol_core.protocol_name', 'analysis_protocol.protocol_core.protocol_description', 'analysis_protocol.protocol_core.publication_doi', 'analysis_protocol.protocol_core.protocols_io_doi', 'analysis_protocol.protocol_core.document', 'analysis_protocol.protocol_type.text', 'analysis_protocol.protocol_type.ontology', 'analysis_protocol.protocol_type.ontology_label', 'analysis_protocol.computational_method', 'protocol.protocol_core.protocol_id', 'protocol.protocol_core.protocol_name', 'protocol.protocol_core.protocol_description', 'protocol.protocol_core.publication_doi', 'protocol.protocol_core.protocols_io_doi', 'protocol.protocol_core.document', 'protocol.protocol_type.text', 'protocol.protocol_type.ontology', 'protocol.protocol_type.ontology_label', 'project.project_core.project_short_name', 'project.project_core.project_title', 'project.project_core.project_description', 'project.contributors.contact_name', 'project.contributors.email', 'project.contributors.phone', 'project.contributors.institution', 'project.contributors.laboratory', 'project.contributors.address', 'project.contributors.country', 'project.contributors.corresponding_contributor', 'project.contributors.project_role', 'project.contributors.orcid_id', 'project.supplementary_links', 'project.publications.authors', 'project.publications.publication_title', 'project.publications.doi', 'project.publications.pmid', 'project.publications.publication_url', 'project.insdc_project_accessions', 'project.geo_series_accessions', 'project.array_express_accessions', 'project.insdc_study_accessions', 'project.funders.grant_title', 'project.funders.grant_id', 'project.funders.organization', 'analysis_process.process_core.process_id', 'analysis_process.process_core.process_name', 'analysis_process.process_core.process_description', 'analysis_process.process_core.process_location', 'analysis_process.process_core.operator', 'analysis_process.process_type.text', 'analysis_process.process_type.ontology', 'analysis_process.process_type.ontology_label', 'analysis_process.inputs.checksum', 'analysis_process.inputs.parameter_name', 'analysis_process.inputs.parameter_value', 'analysis_process.tasks.disk_size', 'analysis_process.tasks.task_name', 'analysis_process.tasks.zone', 'analysis_process.tasks.log_err', 'analysis_process.tasks.start_time', 'analysis_process.tasks.cpus', 'analysis_process.tasks.log_out', 'analysis_process.tasks.stop_time', 'analysis_process.tasks.memory', 'analysis_process.tasks.docker_image', 'analysis_process.input_bundles', 'analysis_process.timestamp_start_utc', 'analysis_process.timestamp_stop_utc', 'analysis_process.analysis_run_type', 'analysis_process.reference_bundle', 'process.process_core.process_id', 'process.process_core.process_name', 'process.process_core.process_description', 'process.process_core.process_location', 'process.process_core.operator', 'process.length_of_time', 'process.length_of_time_unit.text', 'process.length_of_time_unit.ontology', 'process.length_of_time_unit.ontology_label', 'process.process_type.text', 'process.process_type.ontology', 'process.process_type.ontology_label', 'process.deviation_from_protocol', 'process.insdc_experiment.insdc_experiment_accession', 'supplementary_file.file_core.file_name', 'supplementary_file.file_core.file_format', 'supplementary_file.file_core.checksum', 'supplementary_file.file_description', 'sequence_file.file_core.file_name', 'sequence_file.file_core.file_format', 'sequence_file.file_core.checksum', 'sequence_file.read_index', 'sequence_file.lane_index', 'sequence_file.read_length', 'sequence_file.insdc_run_accessions', 'sequence_file.library_prep_id', 'reference_file.file_core.file_name', 'reference_file.file_core.file_format', 'reference_file.file_core.checksum', 'reference_file.ncbi_taxon_id', 'reference_file.genus_species.text', 'reference_file.genus_species.ontology', 'reference_file.genus_species.ontology_label', 'reference_file.reference_type', 'reference_file.assembly_type', 'reference_file.reference_version', 'protocol_file.file_core.file_name', 'protocol_file.file_core.file_format', 'protocol_file.file_core.checksum', 'image_file.file_core.file_name', 'image_file.file_core.file_format', 'image_file.file_core.checksum', 'analysis_file.file_core.file_name', 'analysis_file.file_core.file_format', 'analysis_file.file_core.checksum', 'specimen_from_organism.biomaterial_core.biomaterial_id', 'specimen_from_organism.biomaterial_core.biomaterial_name', 'specimen_from_organism.biomaterial_core.biomaterial_description', 'specimen_from_organism.biomaterial_core.ncbi_taxon_id', 'specimen_from_organism.biomaterial_core.genotype', 'specimen_from_organism.biomaterial_core.supplementary_files', 'specimen_from_organism.biomaterial_core.biosd_biomaterial', 'specimen_from_organism.biomaterial_core.insdc_biomaterial', 'specimen_from_organism.genus_species.text', 'specimen_from_organism.genus_species.ontology', 'specimen_from_organism.genus_species.ontology_label', 'specimen_from_organism.organ.text', 'specimen_from_organism.organ.ontology', 'specimen_from_organism.organ.ontology_label', 'specimen_from_organism.organ_parts.text', 'specimen_from_organism.organ_parts.ontology', 'specimen_from_organism.organ_parts.ontology_label', 'specimen_from_organism.diseases.text', 'specimen_from_organism.diseases.ontology', 'specimen_from_organism.diseases.ontology_label', 'specimen_from_organism.state_of_specimen.autolysis_score', 'specimen_from_organism.state_of_specimen.gross_description', 'specimen_from_organism.state_of_specimen.gross_images', 'specimen_from_organism.state_of_specimen.ischemic_temperature', 'specimen_from_organism.state_of_specimen.ischemic_time', 'specimen_from_organism.state_of_specimen.microscopic_description', 'specimen_from_organism.state_of_specimen.microscopic_images', 'specimen_from_organism.state_of_specimen.postmortem_interval', 'specimen_from_organism.preservation_storage.storage_method', 'specimen_from_organism.preservation_storage.storage_time', 'specimen_from_organism.preservation_storage.storage_time_unit.text', 'specimen_from_organism.preservation_storage.storage_time_unit.ontology', 'specimen_from_organism.preservation_storage.storage_time_unit.ontology_label', 'specimen_from_organism.preservation_storage.preservation_method', 'specimen_from_organism.collection_time', 'specimen_from_organism.purchased_specimen.retail_name', 'specimen_from_organism.purchased_specimen.catalog_number', 'specimen_from_organism.purchased_specimen.manufacturer', 'specimen_from_organism.purchased_specimen.lot_number', 'specimen_from_organism.purchased_specimen.expiry_date', 'specimen_from_organism.purchased_specimen.kit_titer', 'organoid.biomaterial_core.biomaterial_id', 'organoid.biomaterial_core.biomaterial_name', 'organoid.biomaterial_core.biomaterial_description', 'organoid.biomaterial_core.ncbi_taxon_id', 'organoid.biomaterial_core.genotype', 'organoid.biomaterial_core.supplementary_files', 'organoid.biomaterial_core.biosd_biomaterial', 'organoid.biomaterial_core.insdc_biomaterial', 'organoid.genus_species.text', 'organoid.genus_species.ontology', 'organoid.genus_species.ontology_label', 'organoid.model_organ.text', 'organoid.model_organ.ontology', 'organoid.model_organ.ontology_label', 'organoid.model_organ_part.text', 'organoid.model_organ_part.ontology', 'organoid.model_organ_part.ontology_label', 'organoid.age', 'organoid.age_unit.text', 'organoid.age_unit.ontology', 'organoid.age_unit.ontology_label', 'organoid.size', 'organoid.size_unit.text', 'organoid.size_unit.ontology', 'organoid.size_unit.ontology_label', 'organoid.morphology', 'organoid.embedded_in_matrigel', 'organoid.growth_environment', 'organoid.input_aggregate_cell_count', 'organoid.stored_oxygen_levels', 'imaged_specimen.biomaterial_core.biomaterial_id', 'imaged_specimen.biomaterial_core.biomaterial_name', 'imaged_specimen.biomaterial_core.biomaterial_description', 'imaged_specimen.biomaterial_core.ncbi_taxon_id', 'imaged_specimen.biomaterial_core.genotype', 'imaged_specimen.biomaterial_core.supplementary_files', 'imaged_specimen.biomaterial_core.biosd_biomaterial', 'imaged_specimen.biomaterial_core.insdc_biomaterial', 'imaged_specimen.overview_images', 'imaged_specimen.slice_thickness', 'imaged_specimen.internal_anatomical_structures.text', 'imaged_specimen.internal_anatomical_structures.ontology', 'imaged_specimen.internal_anatomical_structures.ontology_label', 'donor_organism.biomaterial_core.biomaterial_id', 'donor_organism.biomaterial_core.biomaterial_name', 'donor_organism.biomaterial_core.biomaterial_description', 'donor_organism.biomaterial_core.ncbi_taxon_id', 'donor_organism.biomaterial_core.genotype', 'donor_organism.biomaterial_core.supplementary_files', 'donor_organism.biomaterial_core.biosd_biomaterial', 'donor_organism.biomaterial_core.insdc_biomaterial', 'donor_organism.human_specific.body_mass_index', 'donor_organism.human_specific.ethnicity.text', 'donor_organism.human_specific.ethnicity.ontology', 'donor_organism.human_specific.ethnicity.ontology_label', 'donor_organism.mouse_specific.strain.text', 'donor_organism.mouse_specific.strain.ontology', 'donor_organism.mouse_specific.strain.ontology_label', 'donor_organism.genus_species.text', 'donor_organism.genus_species.ontology', 'donor_organism.genus_species.ontology_label', 'donor_organism.sex', 'donor_organism.is_living', 'donor_organism.organism_age', 'donor_organism.organism_age_unit.text', 'donor_organism.organism_age_unit.ontology', 'donor_organism.organism_age_unit.ontology_label', 'donor_organism.development_stage.text', 'donor_organism.development_stage.ontology', 'donor_organism.development_stage.ontology_label', 'donor_organism.diseases.text', 'donor_organism.diseases.ontology', 'donor_organism.diseases.ontology_label', 'donor_organism.death.cause_of_death', 'donor_organism.death.cold_perfused', 'donor_organism.death.days_on_ventilator', 'donor_organism.death.hardy_scale', 'donor_organism.death.time_of_death', 'donor_organism.death.organ_donation_death_type', 'donor_organism.death.normothermic_regional_perfusion', 'donor_organism.familial_relationships.parent', 'donor_organism.familial_relationships.child', 'donor_organism.familial_relationships.sibling', 'donor_organism.medical_history.alcohol_history', 'donor_organism.medical_history.medication', 'donor_organism.medical_history.smoking_history', 'donor_organism.medical_history.nutritional_state', 'donor_organism.medical_history.test_results', 'donor_organism.medical_history.treatment', 'donor_organism.gestational_age', 'donor_organism.gestational_age_unit.text', 'donor_organism.gestational_age_unit.ontology', 'donor_organism.gestational_age_unit.ontology_label', 'donor_organism.height', 'donor_organism.height_unit.text', 'donor_organism.height_unit.ontology', 'donor_organism.height_unit.ontology_label', 'donor_organism.weight', 'donor_organism.weight_unit.text', 'donor_organism.weight_unit.ontology', 'donor_organism.weight_unit.ontology_label', 'donor_organism.timecourse.value', 'donor_organism.timecourse.unit.text', 'donor_organism.timecourse.unit.ontology', 'donor_organism.timecourse.unit.ontology_label', 'donor_organism.timecourse.relevance', 'cell_suspension.biomaterial_core.biomaterial_id', 'cell_suspension.biomaterial_core.biomaterial_name', 'cell_suspension.biomaterial_core.biomaterial_description', 'cell_suspension.biomaterial_core.ncbi_taxon_id', 'cell_suspension.biomaterial_core.genotype', 'cell_suspension.biomaterial_core.supplementary_files', 'cell_suspension.biomaterial_core.biosd_biomaterial', 'cell_suspension.biomaterial_core.insdc_biomaterial', 'cell_suspension.cell_morphology.cell_morphology', 'cell_suspension.cell_morphology.cell_size', 'cell_suspension.cell_morphology.cell_size_unit.text', 'cell_suspension.cell_morphology.cell_size_unit.ontology', 'cell_suspension.cell_morphology.cell_size_unit.ontology_label', 'cell_suspension.cell_morphology.percent_cell_viability', 'cell_suspension.cell_morphology.cell_viability_method', 'cell_suspension.cell_morphology.cell_viability_result', 'cell_suspension.cell_morphology.percent_necrosis', 'cell_suspension.growth_conditions.passage_number', 'cell_suspension.growth_conditions.growth_medium', 'cell_suspension.growth_conditions.culture_environment', 'cell_suspension.growth_conditions.mycoplasma_testing_method', 'cell_suspension.growth_conditions.mycoplasma_testing_results', 'cell_suspension.growth_conditions.drug_treatment', 'cell_suspension.growth_conditions.feeder_layer_type', 'cell_suspension.genus_species.text', 'cell_suspension.genus_species.ontology', 'cell_suspension.genus_species.ontology_label', 'cell_suspension.selected_cell_type.text', 'cell_suspension.selected_cell_type.ontology', 'cell_suspension.selected_cell_type.ontology_label', 'cell_suspension.estimated_cell_count', 'cell_suspension.plate_based_sequencing.plate_label', 'cell_suspension.plate_based_sequencing.well_label', 'cell_suspension.plate_based_sequencing.well_quality', 'cell_suspension.timecourse.value', 'cell_suspension.timecourse.unit.text', 'cell_suspension.timecourse.unit.ontology', 'cell_suspension.timecourse.unit.ontology_label', 'cell_suspension.timecourse.relevance', 'cell_line.biomaterial_core.biomaterial_id', 'cell_line.biomaterial_core.biomaterial_name', 'cell_line.biomaterial_core.biomaterial_description', 'cell_line.biomaterial_core.ncbi_taxon_id', 'cell_line.biomaterial_core.genotype', 'cell_line.biomaterial_core.supplementary_files', 'cell_line.biomaterial_core.biosd_biomaterial', 'cell_line.biomaterial_core.insdc_biomaterial', 'cell_line.supplier', 'cell_line.catalog_number', 'cell_line.lot_number', 'cell_line.catalog_url', 'cell_line.cell_cycle.text', 'cell_line.cell_cycle.ontology', 'cell_line.cell_cycle.ontology_label', 'cell_line.cell_line_type', 'cell_line.model_organ.text', 'cell_line.model_organ.ontology', 'cell_line.model_organ.ontology_label', 'cell_line.cell_morphology.cell_morphology', 'cell_line.cell_morphology.cell_size', 'cell_line.cell_morphology.cell_size_unit.text', 'cell_line.cell_morphology.cell_size_unit.ontology', 'cell_line.cell_morphology.cell_size_unit.ontology_label', 'cell_line.cell_morphology.percent_cell_viability', 'cell_line.cell_morphology.cell_viability_method', 'cell_line.cell_morphology.cell_viability_result', 'cell_line.cell_morphology.percent_necrosis', 'cell_line.growth_conditions.passage_number', 'cell_line.growth_conditions.growth_medium', 'cell_line.growth_conditions.culture_environment', 'cell_line.growth_conditions.mycoplasma_testing_method', 'cell_line.growth_conditions.mycoplasma_testing_results', 'cell_line.growth_conditions.drug_treatment', 'cell_line.growth_conditions.feeder_layer_type', 'cell_line.confluency', 'cell_line.cell_type.text', 'cell_line.cell_type.ontology', 'cell_line.cell_type.ontology_label', 'cell_line.karyotype', 'cell_line.tissue.text', 'cell_line.tissue.ontology', 'cell_line.tissue.ontology_label', 'cell_line.date_established', 'cell_line.disease.text', 'cell_line.disease.ontology', 'cell_line.disease.ontology_label', 'cell_line.genus_species.text', 'cell_line.genus_species.ontology', 'cell_line.genus_species.ontology_label', 'cell_line.publication.authors', 'cell_line.publication.publication_title', 'cell_line.publication.doi', 'cell_line.publication.pmid', 'cell_line.publication.publication_url']\n"
     ]
    }
   ],
   "source": [
    "from ingest.template.schema_template import SchemaTemplate\n",
    "import yaml\n",
    "\n",
    "template = SchemaTemplate()\n",
    "attribute_yaml = template.yaml_dump()\n",
    "\n",
    "attribute_list = yaml.load(attribute_yaml).get('tabs')\n",
    "full_set = set()\n",
    "full_list = []\n",
    "for tab in attribute_list:\n",
    "    for x in list(tab.values()):\n",
    "        for attribute in x.get('columns'):\n",
    "            full_set.add(attribute)\n",
    "            full_list.append(attribute)\n",
    "print('{} out of {} attributes are unique'.format(len(full_set), len(full_list)))\n",
    "print(full_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is a temp cell to add some out of date attributes that are present in the cbeta\n",
    "\n",
    "full_list = full_list + ['dissociation_protocol.method.ontology_label',\n",
    "'library_preparation_protocol.library_construction_method.text',\n",
    "'sequencing_protocol.method.text',\n",
    "'dissociation_protocol.method.text',\n",
    "'sequencing_protocol.method.ontology',\n",
    "'specimen_from_organism.organ_parts.text',\n",
    "'specimen_from_organism.organ_parts.ontology_label',\n",
    "'library_preparation_protocol.library_construction_method.ontology_label',\n",
    "'specimen_from_organism.organ_parts.ontology',\n",
    "'sequencing_protocol.method.ontology_label',\n",
    "'library_preparation_protocol.library_construction_method.ontology',\n",
    "'sequence_file.insdc_run_accessions',\n",
    "'cell_suspension.estimated_cell_count',\n",
    "'donor_organism.human_specific.ethnicity.text',\n",
    "'project.insdc_project_accessions',\n",
    "'project.geo_series_accessions',\n",
    "'project.funders.grant_id',\n",
    "'process.insdc_experiment.insdc_experiment_accession',\n",
    "'cell_suspension.plate_based_sequencing.well_quality',\n",
    "'collection_protocol.method.ontology',\n",
    "'project.insdc_study_accessions',\n",
    "'collection_protocol.method.text',\n",
    "'collection_protocol.method.ontology_label',\n",
    "'cell_suspension.plate_based_sequencing.plate_label',\n",
    "'project.array_express_accessions',\n",
    "'cell_suspension.plate_based_sequencing.well_quality',\n",
    "'donor_organism.familial_relationships.child',\n",
    "'donor_organism.familial_relationships.parent',\n",
    "'cell_line.cell_morphology.cell_size_unit.text',\n",
    "'cell_suspension.cell_morphology.cell_size_unit.text',\n",
    "'ipsc_induction_protocol.protocol_reagents.catalog_number',\n",
    "'ipsc_induction_protocol.protocol_reagents.manufacturer',\n",
    "'ipsc_induction_protocol.protocol_reagents.expiry_date',\n",
    "'ipsc_induction_protocol.protocol_reagents.lot_number',\n",
    "'ipsc_induction_protocol.protocol_reagents.retail_name',\n",
    "'donor_organism.mouse_specific.strain.text',\n",
    "'dissociation_protocol.method.ontology']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2a.\n",
    "\n",
    "`initial_q` is the initial Elastic Search query we will use to get the bundles we will feed to the attribute counter in the next step. \n",
    "\n",
    "The top example contains a `must_not` term that will exclude datasets taken from the exclusion list [here](https://github.com/DataBiosphere/azul/blob/c0937a7dcbd6f70f6d28ce004f4005c8eac0d9dc/src/azul/project/hca/__init__.py#L90).\n",
    "\n",
    "**(Reccomended) The bottom example contains a `must` term that will only include projects. These datasets were taben from [here](https://docs.google.com/spreadsheets/d/155ACGfrgobXy9DI_GduYyHCpkCyFDIa6FvgH8keBdug/edit?ts=5c084ebb#gid=287685589)**\n",
    "\n",
    "You can also add `must` so this query only matches specific attributes or specific values. For example you may only want to match human smart seq 2 datasets. For this you could use queries similar to [this one](https://github.com/HumanCellAtlas/lira/blob/master/scripts/vx_queries/smartseq2-query.json) used by our internal analysis pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_q = {\n",
    "    \"query\": {\n",
    "        \"bool\": {\n",
    "            \"must_not\": [\n",
    "                {\n",
    "                    \"terms\": {\n",
    "                        \"files.project_json.provenance.document_id\": [\n",
    "                            \"1630e3dc-5501-4faf-9726-2e2c0b4da6d7\",\n",
    "                            \"fd1d163d-d6a7-41cd-b3bc-9d77ba9a36fe\",\n",
    "                            \"2a0faf83-e342-4b1c-bb9b-cf1d1147f3bb\",\n",
    "                            \"cf8439db-fcc9-44a8-b66f-8ffbf729bffa\",\n",
    "                            \"6b9f514d-d738-403f-a9c2-62580bbe5c83\",\n",
    "                            \"311d013c-01e4-42c0-9c2d-25472afa9cbc\",\n",
    "                            \"d237ed6a-3a7f-4a91-b300-b070888a8542\",\n",
    "                            \"e6cc0b02-2125-4faa-9903-a9025a62efec\",\n",
    "                            \"e4dbcb98-0562-4071-8bea-5e8de5f3c147\",\n",
    "                            \"e79e9284-c337-4dfd-853d-66fa3facfbbd\",\n",
    "                            \"560cd061-9165-4699-bc6e-8253e164c079\",\n",
    "                            \"e83fda0e-6515-4f13-82cb-a5860ecfc2d4\",\n",
    "                            \"9a60e8c2-32ea-4586-bc1f-7ee58f462b07\",\n",
    "                            \"71a6e049-4846-4c2a-8823-cc193c573efc\",\n",
    "                            \"4b5a2268-507c-46e6-bab0-3efb30145e85\",\n",
    "                            \"364ebb73-652e-4d32-8938-1c922d0b2584\",\n",
    "                            \"11f5d59b-0e2c-4f01-85ac-8d8dd3db53be\",\n",
    "                            \"c1996526-6466-40ff-820f-dad4d63492ec\",\n",
    "                            \"c281dedc-e838-4464-bf51-1cc4efae3fb9\",\n",
    "                            \"40afcf6b-422a-47ba-ba7a-33678c949b5c\",\n",
    "                            \"71a6e049-4846-4c2a-8823-cc193c573efc\",\n",
    "                            \"9a60e8c2-32ea-4586-bc1f-7ee58f462b07\",\n",
    "                            \"0facfacd-5b0c-4228-8be5-37aa1f3a269d\",\n",
    "                            \"76c209df-42bf-41dc-a5f5-3d27193ca7a6\",\n",
    "                            \"bb409c34-bb87-4ed2-adaf-6d1ef10610b5\",\n",
    "                            \"1a6b5e5d-914f-4dd6-8817-a1f9b7f364d5\",\n",
    "                            \"dd401943-1059-4b2d-b187-7a9e11822f95\"\n",
    "                        ]\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_q = {\n",
    "    \"query\": {\n",
    "        \"bool\": {\n",
    "            \"must\": [\n",
    "                {\n",
    "                    \"terms\": {\n",
    "                        \"files.project_json.provenance.document_id\": [\n",
    "                            \"d96c2451-6e22-441f-a3e6-70fd0878bb1b\",\n",
    "                            \"e8642221-4c2c-4fd7-b926-a68bce363c88\",\n",
    "                            \"209e6402-d854-49ea-815f-421dae5e3f4d\",\n",
    "                            \"29f53b7e-071b-44b5-998a-0ae70d0229a4\",\n",
    "                            \"f8880be0-210c-4aa3-9348-f5a423e07421\",\n",
    "                            \"aabbec1a-1215-43e1-8e42-6489af25c12c\",\n",
    "                            \"179bf9e6-5b33-4c5b-ae26-96c7270976b8\",\n",
    "                            \"1a0f98b8-746a-489d-8af9-d5c657482aab\",\n",
    "                            \"f396fa53-2a2d-4b8a-ad18-03bf4bd46833\",\n",
    "                            \"0ec2b05f-ddbe-4e5a-b30f-e81f4b1e330c\",\n",
    "                            \"32eb86db-6842-480f-a49a-a2b0161ed35a\",\n",
    "                            \"0c7bbbce-3c70-4d6b-a443-1b92c1f205c8\",\n",
    "                            \"34ec62a2-9643-430d-b41a-1e342bd615fc\",\n",
    "                            \"c765e3f9-7cfc-4501-8832-79e5f7abd321\",\n",
    "                            \"ff481f29-3d0b-4533-9de2-a760c61c162d\",\n",
    "                            \"5f256182-5dfc-4070-8404-f6fa71d37c73\"\n",
    "                        ]\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2b.\n",
    "\n",
    "Making a bundle generator from the DSS. This geenrator will iterate through all the matches but it has optimisation behind the scenes to balance payload per request and number of requests for you.\n",
    "\n",
    "You'll need to `pip install hca` on your machine to use these modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hca.dss\n",
    "from hca.dss import DSSClient\n",
    "\n",
    "dss_client = DSSClient(swagger_url= \"https://dss.data.humancellatlas.org/v1/swagger.json\")\n",
    "bundle_generator = dss_client.post_search.iterate(replica=\"aws\", es_query=initial_q, output_format=\"raw\")\n",
    "total_hits = dss_client.post_search(replica=\"aws\", es_query=initial_q, output_format=\"raw\").get('total_hits')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.\n",
    "\n",
    "In section 3 we will iterate through bundles captured in section 2. For each bundle we will count attributes from the list generated in section 1 then convert these counts into a per project count rather than just per bundle.\n",
    "\n",
    "**As the iterator contains all the bundles in the data store, this counting can take 1-2h** If you want to quickly test the script adjust the `LIMIT_RUN` number to limit the number of bundles you want to process. If you set this to `False` all the bundles that match the query will be processed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c757b24419ae4c13ae48e6a1bedfa0f1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=42254), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MISSING 0 BUNDLES\n",
      "DataFrame contains 590 rows/attributes and 42254 columns/bundles\n"
     ]
    }
   ],
   "source": [
    "LIMIT_RUN = False\n",
    "# LIMIT_RUN = 1000\n",
    "\n",
    "\n",
    "import sys # temp remove now\n",
    "import json # temp remove now\n",
    "import pandas as pd\n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "\n",
    "temp_counter = 0 # remove for full run\n",
    "attribute_present_all = {}\n",
    "project_mapping = {}\n",
    "for bundle in tqdm(bundle_generator, total=total_hits, unit='bundle'):\n",
    "    \n",
    "    bundle_url = bundle.get('bundle_url')\n",
    "    bundle_meta = bundle.get('metadata').get('files')\n",
    "    attribute_present = {}\n",
    "    missing_bundle_counter = 0\n",
    "    if bundle_meta:\n",
    "\n",
    "        # make project uuid to bundle uuid mapping dict\n",
    "        try:\n",
    "            project_uuid = bundle_meta.get('project_json')[0].get('provenance').get('document_id')\n",
    "        except KeyError:\n",
    "            project_uuid = bundle_meta.get('project_json').get('hca_ingest').get('document_id')\n",
    "            \n",
    "        if project_uuid in project_mapping:\n",
    "            old_value = project_mapping.get(project_uuid)\n",
    "            new_value = old_value + [bundle_url]\n",
    "            project_mapping[project_uuid] = new_value\n",
    "            \n",
    "        else:\n",
    "            project_mapping[project_uuid] = [bundle_url]\n",
    "   \n",
    "        # search for all attributes in the bundle\n",
    "        \n",
    "        for attribute in full_list:\n",
    "            nesting = attribute.split('.')\n",
    "            nesting[0] = nesting[0] + '_json'\n",
    "            nested_get = bundle_meta.get(nesting[0])\n",
    "            \n",
    "            if nested_get:\n",
    "                for metadata_entity in nested_get:\n",
    "                    for idx, level in enumerate(nesting[1:]):\n",
    "                    \n",
    "                        try:\n",
    "                            if attribute_present.get(attribute) == True: # if attribute already found in bundle\n",
    "                                if idx == len(nesting[1:]):\n",
    "                                    break\n",
    "                            metadata_entity = metadata_entity.get(level)\n",
    "                            attribute_present[attribute] = True\n",
    "                            \n",
    "                            \n",
    "                        except AttributeError:\n",
    "                            attribute_present[attribute] = False\n",
    "  \n",
    "            else:\n",
    "                attribute_present[attribute] = False\n",
    "    else:\n",
    "        missing_bundle_counter += 1\n",
    "    attribute_present_all[bundle_url] = (attribute_present)\n",
    "    temp_counter += 1 # remove for full run\n",
    "    if temp_counter == LIMIT_RUN:\n",
    "        break\n",
    "        \n",
    "    \n",
    "print('MISSING {} BUNDLES'.format(missing_bundle_counter))\n",
    "\n",
    "df = pd.DataFrame(attribute_present_all)\n",
    "df.to_csv('temp_df.csv') # save out for inspection\n",
    "print('DataFrame contains {} rows/attributes and {} columns/bundles'.format(df.shape[0], df.shape[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3b.\n",
    "\n",
    "This section counts up the results from the dataframe created above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c16902ec516042608b3217adb9ed4bf1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=627), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "bundle_counts = df.count(axis=1, level=None, numeric_only=False)\n",
    "# print(bundle_counts)\n",
    "# bundle_counts.to_csv('bundle_counts.csv')\n",
    "\n",
    "\n",
    "bundle_counts = {}\n",
    "\n",
    "for attribute in tqdm(full_list, unit='attribute'):\n",
    "    value_counts = df.T[attribute].value_counts()\n",
    "    true_counts = value_counts.get(True)\n",
    "    false_counts = value_counts.get(False)\n",
    "    \n",
    "    bundle_counts[attribute] = {'True': true_counts, 'False' : false_counts}\n",
    "    \n",
    "# print(bundle_counts) # a dictionary containing counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3c.\n",
    "\n",
    "In the previous section we attribute usage in bundles but ideally we want to know the counts on a project basis. Using the bundle to project mapping `project_mapping` we created earler, here we go..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "20e9dc0c9a194872bc18887b17fd331d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=627), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute NOT used (no. projects)</th>\n",
       "      <th>Attribute usage (no. projects)</th>\n",
       "      <th>Schema Name</th>\n",
       "      <th>Schema usage (no. project)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocol_id</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocol_name</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocol_description</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.publication_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocols_io_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.document</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.instrument_manufacturer_model.text</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.instrument_manufacturer_model.ontology</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.instrument_manufacturer_model.ontology_label</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.local_machine_name</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.paired_end</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.method.text</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.method.ontology</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.method.ontology_label</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.fastq_method</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.fastq_method_version</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.pooled_channels</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.drop_uniformity</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocol_id</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocol_name</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocol_description</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.publication_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocols_io_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.document</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.barcode_read</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.barcode_offset</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.barcode_length</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.white_list_file</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.input_nucleic_acid_molecule.text</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.input_nucleic_acid_molecule.ontology</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_morphology.cell_viability_method</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_morphology.cell_viability_result</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_morphology.percent_necrosis</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.passage_number</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.growth_medium</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.culture_environment</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.mycoplasma_testing_method</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.mycoplasma_testing_results</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.drug_treatment</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.feeder_layer_type</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.confluency</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_type.text</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_type.ontology</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_type.ontology_label</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.karyotype</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.tissue.text</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.tissue.ontology</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.tissue.ontology_label</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.date_established</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.disease.text</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.disease.ontology</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.disease.ontology_label</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.genus_species.text</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.genus_species.ontology</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.genus_species.ontology_label</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.authors</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.publication_title</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.doi</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.pmid</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.publication_url</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>590 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Attribute NOT used (no. projects)  \\\n",
       "index                                                                                   \n",
       "sequencing_protocol.protocol_core.protocol_id                                     0.0   \n",
       "sequencing_protocol.protocol_core.protocol_name                                   0.0   \n",
       "sequencing_protocol.protocol_core.protocol_desc...                                0.0   \n",
       "sequencing_protocol.protocol_core.publication_doi                                 0.0   \n",
       "sequencing_protocol.protocol_core.protocols_io_doi                                0.0   \n",
       "sequencing_protocol.protocol_core.document                                        0.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                                0.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                                0.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                                0.0   \n",
       "sequencing_protocol.local_machine_name                                            0.0   \n",
       "sequencing_protocol.paired_end                                                    0.0   \n",
       "sequencing_protocol.method.text                                                  15.0   \n",
       "sequencing_protocol.method.ontology                                              15.0   \n",
       "sequencing_protocol.method.ontology_label                                        15.0   \n",
       "sequencing_protocol.10x.fastq_method                                             15.0   \n",
       "sequencing_protocol.10x.fastq_method_version                                     15.0   \n",
       "sequencing_protocol.10x.pooled_channels                                          15.0   \n",
       "sequencing_protocol.10x.drop_uniformity                                          15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.publ...                                0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.docu...                                0.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                                5.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                                5.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                                5.0   \n",
       "library_preparation_protocol.cell_barcode.white...                                5.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                                0.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                                0.0   \n",
       "...                                                                               ...   \n",
       "cell_line.cell_morphology.cell_viability_method                                  14.0   \n",
       "cell_line.cell_morphology.cell_viability_result                                  14.0   \n",
       "cell_line.cell_morphology.percent_necrosis                                       14.0   \n",
       "cell_line.growth_conditions.passage_number                                       13.0   \n",
       "cell_line.growth_conditions.growth_medium                                        13.0   \n",
       "cell_line.growth_conditions.culture_environment                                  13.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                               13.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                               13.0   \n",
       "cell_line.growth_conditions.drug_treatment                                       13.0   \n",
       "cell_line.growth_conditions.feeder_layer_type                                    13.0   \n",
       "cell_line.confluency                                                             11.0   \n",
       "cell_line.cell_type.text                                                         11.0   \n",
       "cell_line.cell_type.ontology                                                     11.0   \n",
       "cell_line.cell_type.ontology_label                                               11.0   \n",
       "cell_line.karyotype                                                              11.0   \n",
       "cell_line.tissue.text                                                            15.0   \n",
       "cell_line.tissue.ontology                                                        15.0   \n",
       "cell_line.tissue.ontology_label                                                  15.0   \n",
       "cell_line.date_established                                                       11.0   \n",
       "cell_line.disease.text                                                           13.0   \n",
       "cell_line.disease.ontology                                                       13.0   \n",
       "cell_line.disease.ontology_label                                                 13.0   \n",
       "cell_line.genus_species.text                                                     15.0   \n",
       "cell_line.genus_species.ontology                                                 15.0   \n",
       "cell_line.genus_species.ontology_label                                           15.0   \n",
       "cell_line.publication.authors                                                    14.0   \n",
       "cell_line.publication.publication_title                                          14.0   \n",
       "cell_line.publication.doi                                                        14.0   \n",
       "cell_line.publication.pmid                                                       14.0   \n",
       "cell_line.publication.publication_url                                            14.0   \n",
       "\n",
       "                                                    Attribute usage (no. projects)  \\\n",
       "index                                                                                \n",
       "sequencing_protocol.protocol_core.protocol_id                                 15.0   \n",
       "sequencing_protocol.protocol_core.protocol_name                               15.0   \n",
       "sequencing_protocol.protocol_core.protocol_desc...                            15.0   \n",
       "sequencing_protocol.protocol_core.publication_doi                             15.0   \n",
       "sequencing_protocol.protocol_core.protocols_io_doi                            15.0   \n",
       "sequencing_protocol.protocol_core.document                                    15.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                            15.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                            15.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                            15.0   \n",
       "sequencing_protocol.local_machine_name                                        15.0   \n",
       "sequencing_protocol.paired_end                                                15.0   \n",
       "sequencing_protocol.method.text                                                0.0   \n",
       "sequencing_protocol.method.ontology                                            0.0   \n",
       "sequencing_protocol.method.ontology_label                                      0.0   \n",
       "sequencing_protocol.10x.fastq_method                                           0.0   \n",
       "sequencing_protocol.10x.fastq_method_version                                   0.0   \n",
       "sequencing_protocol.10x.pooled_channels                                        0.0   \n",
       "sequencing_protocol.10x.drop_uniformity                                        0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.publ...                            15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.docu...                            15.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                            10.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                            10.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                            10.0   \n",
       "library_preparation_protocol.cell_barcode.white...                            10.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                            15.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                            15.0   \n",
       "...                                                                            ...   \n",
       "cell_line.cell_morphology.cell_viability_method                                1.0   \n",
       "cell_line.cell_morphology.cell_viability_result                                1.0   \n",
       "cell_line.cell_morphology.percent_necrosis                                     1.0   \n",
       "cell_line.growth_conditions.passage_number                                     2.0   \n",
       "cell_line.growth_conditions.growth_medium                                      2.0   \n",
       "cell_line.growth_conditions.culture_environment                                2.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                             2.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                             2.0   \n",
       "cell_line.growth_conditions.drug_treatment                                     2.0   \n",
       "cell_line.growth_conditions.feeder_layer_type                                  2.0   \n",
       "cell_line.confluency                                                           4.0   \n",
       "cell_line.cell_type.text                                                       4.0   \n",
       "cell_line.cell_type.ontology                                                   4.0   \n",
       "cell_line.cell_type.ontology_label                                             4.0   \n",
       "cell_line.karyotype                                                            4.0   \n",
       "cell_line.tissue.text                                                          0.0   \n",
       "cell_line.tissue.ontology                                                      0.0   \n",
       "cell_line.tissue.ontology_label                                                0.0   \n",
       "cell_line.date_established                                                     4.0   \n",
       "cell_line.disease.text                                                         2.0   \n",
       "cell_line.disease.ontology                                                     2.0   \n",
       "cell_line.disease.ontology_label                                               2.0   \n",
       "cell_line.genus_species.text                                                   0.0   \n",
       "cell_line.genus_species.ontology                                               0.0   \n",
       "cell_line.genus_species.ontology_label                                         0.0   \n",
       "cell_line.publication.authors                                                  1.0   \n",
       "cell_line.publication.publication_title                                        1.0   \n",
       "cell_line.publication.doi                                                      1.0   \n",
       "cell_line.publication.pmid                                                     1.0   \n",
       "cell_line.publication.publication_url                                          1.0   \n",
       "\n",
       "                                                                     Schema Name  \\\n",
       "index                                                                              \n",
       "sequencing_protocol.protocol_core.protocol_id                sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.protocol_name              sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.protocol_desc...           sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.publication_doi            sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.protocols_io_doi           sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.document                   sequencing_protocol   \n",
       "sequencing_protocol.instrument_manufacturer_mod...           sequencing_protocol   \n",
       "sequencing_protocol.instrument_manufacturer_mod...           sequencing_protocol   \n",
       "sequencing_protocol.instrument_manufacturer_mod...           sequencing_protocol   \n",
       "sequencing_protocol.local_machine_name                       sequencing_protocol   \n",
       "sequencing_protocol.paired_end                               sequencing_protocol   \n",
       "sequencing_protocol.method.text                              sequencing_protocol   \n",
       "sequencing_protocol.method.ontology                          sequencing_protocol   \n",
       "sequencing_protocol.method.ontology_label                    sequencing_protocol   \n",
       "sequencing_protocol.10x.fastq_method                         sequencing_protocol   \n",
       "sequencing_protocol.10x.fastq_method_version                 sequencing_protocol   \n",
       "sequencing_protocol.10x.pooled_channels                      sequencing_protocol   \n",
       "sequencing_protocol.10x.drop_uniformity                      sequencing_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.publ...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.docu...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.barco...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.barco...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.barco...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.white...  library_preparation_protocol   \n",
       "library_preparation_protocol.input_nucleic_acid...  library_preparation_protocol   \n",
       "library_preparation_protocol.input_nucleic_acid...  library_preparation_protocol   \n",
       "...                                                                          ...   \n",
       "cell_line.cell_morphology.cell_viability_method                        cell_line   \n",
       "cell_line.cell_morphology.cell_viability_result                        cell_line   \n",
       "cell_line.cell_morphology.percent_necrosis                             cell_line   \n",
       "cell_line.growth_conditions.passage_number                             cell_line   \n",
       "cell_line.growth_conditions.growth_medium                              cell_line   \n",
       "cell_line.growth_conditions.culture_environment                        cell_line   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                     cell_line   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                     cell_line   \n",
       "cell_line.growth_conditions.drug_treatment                             cell_line   \n",
       "cell_line.growth_conditions.feeder_layer_type                          cell_line   \n",
       "cell_line.confluency                                                   cell_line   \n",
       "cell_line.cell_type.text                                               cell_line   \n",
       "cell_line.cell_type.ontology                                           cell_line   \n",
       "cell_line.cell_type.ontology_label                                     cell_line   \n",
       "cell_line.karyotype                                                    cell_line   \n",
       "cell_line.tissue.text                                                  cell_line   \n",
       "cell_line.tissue.ontology                                              cell_line   \n",
       "cell_line.tissue.ontology_label                                        cell_line   \n",
       "cell_line.date_established                                             cell_line   \n",
       "cell_line.disease.text                                                 cell_line   \n",
       "cell_line.disease.ontology                                             cell_line   \n",
       "cell_line.disease.ontology_label                                       cell_line   \n",
       "cell_line.genus_species.text                                           cell_line   \n",
       "cell_line.genus_species.ontology                                       cell_line   \n",
       "cell_line.genus_species.ontology_label                                 cell_line   \n",
       "cell_line.publication.authors                                          cell_line   \n",
       "cell_line.publication.publication_title                                cell_line   \n",
       "cell_line.publication.doi                                              cell_line   \n",
       "cell_line.publication.pmid                                             cell_line   \n",
       "cell_line.publication.publication_url                                  cell_line   \n",
       "\n",
       "                                                    Schema usage (no. project)  \n",
       "index                                                                           \n",
       "sequencing_protocol.protocol_core.protocol_id                             15.0  \n",
       "sequencing_protocol.protocol_core.protocol_name                           15.0  \n",
       "sequencing_protocol.protocol_core.protocol_desc...                        15.0  \n",
       "sequencing_protocol.protocol_core.publication_doi                         15.0  \n",
       "sequencing_protocol.protocol_core.protocols_io_doi                        15.0  \n",
       "sequencing_protocol.protocol_core.document                                15.0  \n",
       "sequencing_protocol.instrument_manufacturer_mod...                        15.0  \n",
       "sequencing_protocol.instrument_manufacturer_mod...                        15.0  \n",
       "sequencing_protocol.instrument_manufacturer_mod...                        15.0  \n",
       "sequencing_protocol.local_machine_name                                    15.0  \n",
       "sequencing_protocol.paired_end                                            15.0  \n",
       "sequencing_protocol.method.text                                           15.0  \n",
       "sequencing_protocol.method.ontology                                       15.0  \n",
       "sequencing_protocol.method.ontology_label                                 15.0  \n",
       "sequencing_protocol.10x.fastq_method                                      15.0  \n",
       "sequencing_protocol.10x.fastq_method_version                              15.0  \n",
       "sequencing_protocol.10x.pooled_channels                                   15.0  \n",
       "sequencing_protocol.10x.drop_uniformity                                   15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.publ...                        15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.docu...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.barco...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.barco...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.barco...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.white...                        15.0  \n",
       "library_preparation_protocol.input_nucleic_acid...                        15.0  \n",
       "library_preparation_protocol.input_nucleic_acid...                        15.0  \n",
       "...                                                                        ...  \n",
       "cell_line.cell_morphology.cell_viability_method                            4.0  \n",
       "cell_line.cell_morphology.cell_viability_result                            4.0  \n",
       "cell_line.cell_morphology.percent_necrosis                                 4.0  \n",
       "cell_line.growth_conditions.passage_number                                 4.0  \n",
       "cell_line.growth_conditions.growth_medium                                  4.0  \n",
       "cell_line.growth_conditions.culture_environment                            4.0  \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                         4.0  \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                         4.0  \n",
       "cell_line.growth_conditions.drug_treatment                                 4.0  \n",
       "cell_line.growth_conditions.feeder_layer_type                              4.0  \n",
       "cell_line.confluency                                                       4.0  \n",
       "cell_line.cell_type.text                                                   4.0  \n",
       "cell_line.cell_type.ontology                                               4.0  \n",
       "cell_line.cell_type.ontology_label                                         4.0  \n",
       "cell_line.karyotype                                                        4.0  \n",
       "cell_line.tissue.text                                                      4.0  \n",
       "cell_line.tissue.ontology                                                  4.0  \n",
       "cell_line.tissue.ontology_label                                            4.0  \n",
       "cell_line.date_established                                                 4.0  \n",
       "cell_line.disease.text                                                     4.0  \n",
       "cell_line.disease.ontology                                                 4.0  \n",
       "cell_line.disease.ontology_label                                           4.0  \n",
       "cell_line.genus_species.text                                               4.0  \n",
       "cell_line.genus_species.ontology                                           4.0  \n",
       "cell_line.genus_species.ontology_label                                     4.0  \n",
       "cell_line.publication.authors                                              4.0  \n",
       "cell_line.publication.publication_title                                    4.0  \n",
       "cell_line.publication.doi                                                  4.0  \n",
       "cell_line.publication.pmid                                                 4.0  \n",
       "cell_line.publication.publication_url                                      4.0  \n",
       "\n",
       "[590 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "project_bool = {}\n",
    "\n",
    "for project_uuid, bundles in project_mapping.items():\n",
    "    project_df = df[bundles]\n",
    "    for attribute in full_list:\n",
    "        attribute_exists = project_df.T[attribute].any()\n",
    "        \n",
    "        if project_uuid in project_bool:\n",
    "            old_value = project_bool.get(project_uuid)\n",
    "            new_value = old_value[attribute] = attribute_exists\n",
    "        else:\n",
    "            project_bool[project_uuid] = {attribute : attribute_exists}\n",
    "        \n",
    "project_df = pd.DataFrame(project_bool)\n",
    "        \n",
    "project_counts = {}\n",
    "\n",
    "for attribute in tqdm(full_list, unit='attribute'):\n",
    "    value_counts = project_df.T[attribute].value_counts()\n",
    "    true_counts = value_counts.get(True)\n",
    "    false_counts = value_counts.get(False)\n",
    "    \n",
    "    project_counts[attribute] = {'True': true_counts, 'False' : false_counts} \n",
    "        \n",
    "# print(project_counts) # a dictionary containing counts\n",
    "project_counts_df = pd.DataFrame(project_counts).T.fillna(value=0)\n",
    "# project_counts_df.to_csv('project_counts.csv')\n",
    "\n",
    "# add columns counting whole schema usage\n",
    "\n",
    "def get_schema(attribute_name):\n",
    "    levels = attribute_name.split('.')\n",
    "    return levels[0]\n",
    "    \n",
    "count_schema = project_counts_df.reset_index()\n",
    "count_schema['schema'] = count_schema['index'].apply(get_schema)\n",
    "count_schema['Schema usage (no. project)'] = count_schema.groupby('schema')['True'].transform('max')\n",
    "count_schema = count_schema.set_index(['index']).rename(index=str, columns={\"True\": \"Attribute usage (no. projects)\", \"False\": \"Attribute NOT used (no. projects)\", \"schema\" : \"Schema Name\"})\n",
    "# count_schema.to_csv('project_counts.csv')\n",
    "count_schema\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute NOT used (no. projects)</th>\n",
       "      <th>Attribute usage (no. projects)</th>\n",
       "      <th>Schema Name</th>\n",
       "      <th>Schema usage (no. project)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocol_id</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocol_name</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocol_description</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.publication_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.protocols_io_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.protocol_core.document</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.instrument_manufacturer_model.text</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.instrument_manufacturer_model.ontology</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.instrument_manufacturer_model.ontology_label</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.local_machine_name</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.paired_end</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.method.text</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.method.ontology</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.method.ontology_label</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.fastq_method</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.fastq_method_version</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.pooled_channels</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_protocol.10x.drop_uniformity</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sequencing_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocol_id</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocol_name</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocol_description</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.publication_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.protocols_io_doi</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.protocol_core.document</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.barcode_read</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.barcode_offset</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.barcode_length</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.cell_barcode.white_list_file</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.input_nucleic_acid_molecule.text</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_preparation_protocol.input_nucleic_acid_molecule.ontology</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>library_preparation_protocol</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_morphology.cell_viability_method</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_morphology.cell_viability_result</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_morphology.percent_necrosis</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.passage_number</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.growth_medium</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.culture_environment</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.mycoplasma_testing_method</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.mycoplasma_testing_results</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.drug_treatment</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.growth_conditions.feeder_layer_type</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.confluency</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_type.text</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_type.ontology</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.cell_type.ontology_label</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.karyotype</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.tissue.text</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.tissue.ontology</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.tissue.ontology_label</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.date_established</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.disease.text</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.disease.ontology</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.disease.ontology_label</th>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.genus_species.text</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.genus_species.ontology</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.genus_species.ontology_label</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.authors</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.publication_title</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.doi</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.pmid</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line.publication.publication_url</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>590 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Attribute NOT used (no. projects)  \\\n",
       "index                                                                                   \n",
       "sequencing_protocol.protocol_core.protocol_id                                     0.0   \n",
       "sequencing_protocol.protocol_core.protocol_name                                   0.0   \n",
       "sequencing_protocol.protocol_core.protocol_desc...                                0.0   \n",
       "sequencing_protocol.protocol_core.publication_doi                                 0.0   \n",
       "sequencing_protocol.protocol_core.protocols_io_doi                                0.0   \n",
       "sequencing_protocol.protocol_core.document                                        0.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                                0.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                                0.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                                0.0   \n",
       "sequencing_protocol.local_machine_name                                            0.0   \n",
       "sequencing_protocol.paired_end                                                    0.0   \n",
       "sequencing_protocol.method.text                                                  15.0   \n",
       "sequencing_protocol.method.ontology                                              15.0   \n",
       "sequencing_protocol.method.ontology_label                                        15.0   \n",
       "sequencing_protocol.10x.fastq_method                                             15.0   \n",
       "sequencing_protocol.10x.fastq_method_version                                     15.0   \n",
       "sequencing_protocol.10x.pooled_channels                                          15.0   \n",
       "sequencing_protocol.10x.drop_uniformity                                          15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.publ...                                0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                                0.0   \n",
       "library_preparation_protocol.protocol_core.docu...                                0.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                                5.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                                5.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                                5.0   \n",
       "library_preparation_protocol.cell_barcode.white...                                5.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                                0.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                                0.0   \n",
       "...                                                                               ...   \n",
       "cell_line.cell_morphology.cell_viability_method                                  14.0   \n",
       "cell_line.cell_morphology.cell_viability_result                                  14.0   \n",
       "cell_line.cell_morphology.percent_necrosis                                       14.0   \n",
       "cell_line.growth_conditions.passage_number                                       13.0   \n",
       "cell_line.growth_conditions.growth_medium                                        13.0   \n",
       "cell_line.growth_conditions.culture_environment                                  13.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                               13.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                               13.0   \n",
       "cell_line.growth_conditions.drug_treatment                                       13.0   \n",
       "cell_line.growth_conditions.feeder_layer_type                                    13.0   \n",
       "cell_line.confluency                                                             11.0   \n",
       "cell_line.cell_type.text                                                         11.0   \n",
       "cell_line.cell_type.ontology                                                     11.0   \n",
       "cell_line.cell_type.ontology_label                                               11.0   \n",
       "cell_line.karyotype                                                              11.0   \n",
       "cell_line.tissue.text                                                            15.0   \n",
       "cell_line.tissue.ontology                                                        15.0   \n",
       "cell_line.tissue.ontology_label                                                  15.0   \n",
       "cell_line.date_established                                                       11.0   \n",
       "cell_line.disease.text                                                           13.0   \n",
       "cell_line.disease.ontology                                                       13.0   \n",
       "cell_line.disease.ontology_label                                                 13.0   \n",
       "cell_line.genus_species.text                                                     15.0   \n",
       "cell_line.genus_species.ontology                                                 15.0   \n",
       "cell_line.genus_species.ontology_label                                           15.0   \n",
       "cell_line.publication.authors                                                    14.0   \n",
       "cell_line.publication.publication_title                                          14.0   \n",
       "cell_line.publication.doi                                                        14.0   \n",
       "cell_line.publication.pmid                                                       14.0   \n",
       "cell_line.publication.publication_url                                            14.0   \n",
       "\n",
       "                                                    Attribute usage (no. projects)  \\\n",
       "index                                                                                \n",
       "sequencing_protocol.protocol_core.protocol_id                                 15.0   \n",
       "sequencing_protocol.protocol_core.protocol_name                               15.0   \n",
       "sequencing_protocol.protocol_core.protocol_desc...                            15.0   \n",
       "sequencing_protocol.protocol_core.publication_doi                             15.0   \n",
       "sequencing_protocol.protocol_core.protocols_io_doi                            15.0   \n",
       "sequencing_protocol.protocol_core.document                                    15.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                            15.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                            15.0   \n",
       "sequencing_protocol.instrument_manufacturer_mod...                            15.0   \n",
       "sequencing_protocol.local_machine_name                                        15.0   \n",
       "sequencing_protocol.paired_end                                                15.0   \n",
       "sequencing_protocol.method.text                                                0.0   \n",
       "sequencing_protocol.method.ontology                                            0.0   \n",
       "sequencing_protocol.method.ontology_label                                      0.0   \n",
       "sequencing_protocol.10x.fastq_method                                           0.0   \n",
       "sequencing_protocol.10x.fastq_method_version                                   0.0   \n",
       "sequencing_protocol.10x.pooled_channels                                        0.0   \n",
       "sequencing_protocol.10x.drop_uniformity                                        0.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.publ...                            15.0   \n",
       "library_preparation_protocol.protocol_core.prot...                            15.0   \n",
       "library_preparation_protocol.protocol_core.docu...                            15.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                            10.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                            10.0   \n",
       "library_preparation_protocol.cell_barcode.barco...                            10.0   \n",
       "library_preparation_protocol.cell_barcode.white...                            10.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                            15.0   \n",
       "library_preparation_protocol.input_nucleic_acid...                            15.0   \n",
       "...                                                                            ...   \n",
       "cell_line.cell_morphology.cell_viability_method                                1.0   \n",
       "cell_line.cell_morphology.cell_viability_result                                1.0   \n",
       "cell_line.cell_morphology.percent_necrosis                                     1.0   \n",
       "cell_line.growth_conditions.passage_number                                     2.0   \n",
       "cell_line.growth_conditions.growth_medium                                      2.0   \n",
       "cell_line.growth_conditions.culture_environment                                2.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                             2.0   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                             2.0   \n",
       "cell_line.growth_conditions.drug_treatment                                     2.0   \n",
       "cell_line.growth_conditions.feeder_layer_type                                  2.0   \n",
       "cell_line.confluency                                                           4.0   \n",
       "cell_line.cell_type.text                                                       4.0   \n",
       "cell_line.cell_type.ontology                                                   4.0   \n",
       "cell_line.cell_type.ontology_label                                             4.0   \n",
       "cell_line.karyotype                                                            4.0   \n",
       "cell_line.tissue.text                                                          0.0   \n",
       "cell_line.tissue.ontology                                                      0.0   \n",
       "cell_line.tissue.ontology_label                                                0.0   \n",
       "cell_line.date_established                                                     4.0   \n",
       "cell_line.disease.text                                                         2.0   \n",
       "cell_line.disease.ontology                                                     2.0   \n",
       "cell_line.disease.ontology_label                                               2.0   \n",
       "cell_line.genus_species.text                                                   0.0   \n",
       "cell_line.genus_species.ontology                                               0.0   \n",
       "cell_line.genus_species.ontology_label                                         0.0   \n",
       "cell_line.publication.authors                                                  1.0   \n",
       "cell_line.publication.publication_title                                        1.0   \n",
       "cell_line.publication.doi                                                      1.0   \n",
       "cell_line.publication.pmid                                                     1.0   \n",
       "cell_line.publication.publication_url                                          1.0   \n",
       "\n",
       "                                                                     Schema Name  \\\n",
       "index                                                                              \n",
       "sequencing_protocol.protocol_core.protocol_id                sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.protocol_name              sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.protocol_desc...           sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.publication_doi            sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.protocols_io_doi           sequencing_protocol   \n",
       "sequencing_protocol.protocol_core.document                   sequencing_protocol   \n",
       "sequencing_protocol.instrument_manufacturer_mod...           sequencing_protocol   \n",
       "sequencing_protocol.instrument_manufacturer_mod...           sequencing_protocol   \n",
       "sequencing_protocol.instrument_manufacturer_mod...           sequencing_protocol   \n",
       "sequencing_protocol.local_machine_name                       sequencing_protocol   \n",
       "sequencing_protocol.paired_end                               sequencing_protocol   \n",
       "sequencing_protocol.method.text                              sequencing_protocol   \n",
       "sequencing_protocol.method.ontology                          sequencing_protocol   \n",
       "sequencing_protocol.method.ontology_label                    sequencing_protocol   \n",
       "sequencing_protocol.10x.fastq_method                         sequencing_protocol   \n",
       "sequencing_protocol.10x.fastq_method_version                 sequencing_protocol   \n",
       "sequencing_protocol.10x.pooled_channels                      sequencing_protocol   \n",
       "sequencing_protocol.10x.drop_uniformity                      sequencing_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.publ...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.prot...  library_preparation_protocol   \n",
       "library_preparation_protocol.protocol_core.docu...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.barco...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.barco...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.barco...  library_preparation_protocol   \n",
       "library_preparation_protocol.cell_barcode.white...  library_preparation_protocol   \n",
       "library_preparation_protocol.input_nucleic_acid...  library_preparation_protocol   \n",
       "library_preparation_protocol.input_nucleic_acid...  library_preparation_protocol   \n",
       "...                                                                          ...   \n",
       "cell_line.cell_morphology.cell_viability_method                        cell_line   \n",
       "cell_line.cell_morphology.cell_viability_result                        cell_line   \n",
       "cell_line.cell_morphology.percent_necrosis                             cell_line   \n",
       "cell_line.growth_conditions.passage_number                             cell_line   \n",
       "cell_line.growth_conditions.growth_medium                              cell_line   \n",
       "cell_line.growth_conditions.culture_environment                        cell_line   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                     cell_line   \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                     cell_line   \n",
       "cell_line.growth_conditions.drug_treatment                             cell_line   \n",
       "cell_line.growth_conditions.feeder_layer_type                          cell_line   \n",
       "cell_line.confluency                                                   cell_line   \n",
       "cell_line.cell_type.text                                               cell_line   \n",
       "cell_line.cell_type.ontology                                           cell_line   \n",
       "cell_line.cell_type.ontology_label                                     cell_line   \n",
       "cell_line.karyotype                                                    cell_line   \n",
       "cell_line.tissue.text                                                  cell_line   \n",
       "cell_line.tissue.ontology                                              cell_line   \n",
       "cell_line.tissue.ontology_label                                        cell_line   \n",
       "cell_line.date_established                                             cell_line   \n",
       "cell_line.disease.text                                                 cell_line   \n",
       "cell_line.disease.ontology                                             cell_line   \n",
       "cell_line.disease.ontology_label                                       cell_line   \n",
       "cell_line.genus_species.text                                           cell_line   \n",
       "cell_line.genus_species.ontology                                       cell_line   \n",
       "cell_line.genus_species.ontology_label                                 cell_line   \n",
       "cell_line.publication.authors                                          cell_line   \n",
       "cell_line.publication.publication_title                                cell_line   \n",
       "cell_line.publication.doi                                              cell_line   \n",
       "cell_line.publication.pmid                                             cell_line   \n",
       "cell_line.publication.publication_url                                  cell_line   \n",
       "\n",
       "                                                    Schema usage (no. project)  \n",
       "index                                                                           \n",
       "sequencing_protocol.protocol_core.protocol_id                             15.0  \n",
       "sequencing_protocol.protocol_core.protocol_name                           15.0  \n",
       "sequencing_protocol.protocol_core.protocol_desc...                        15.0  \n",
       "sequencing_protocol.protocol_core.publication_doi                         15.0  \n",
       "sequencing_protocol.protocol_core.protocols_io_doi                        15.0  \n",
       "sequencing_protocol.protocol_core.document                                15.0  \n",
       "sequencing_protocol.instrument_manufacturer_mod...                        15.0  \n",
       "sequencing_protocol.instrument_manufacturer_mod...                        15.0  \n",
       "sequencing_protocol.instrument_manufacturer_mod...                        15.0  \n",
       "sequencing_protocol.local_machine_name                                    15.0  \n",
       "sequencing_protocol.paired_end                                            15.0  \n",
       "sequencing_protocol.method.text                                           15.0  \n",
       "sequencing_protocol.method.ontology                                       15.0  \n",
       "sequencing_protocol.method.ontology_label                                 15.0  \n",
       "sequencing_protocol.10x.fastq_method                                      15.0  \n",
       "sequencing_protocol.10x.fastq_method_version                              15.0  \n",
       "sequencing_protocol.10x.pooled_channels                                   15.0  \n",
       "sequencing_protocol.10x.drop_uniformity                                   15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.publ...                        15.0  \n",
       "library_preparation_protocol.protocol_core.prot...                        15.0  \n",
       "library_preparation_protocol.protocol_core.docu...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.barco...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.barco...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.barco...                        15.0  \n",
       "library_preparation_protocol.cell_barcode.white...                        15.0  \n",
       "library_preparation_protocol.input_nucleic_acid...                        15.0  \n",
       "library_preparation_protocol.input_nucleic_acid...                        15.0  \n",
       "...                                                                        ...  \n",
       "cell_line.cell_morphology.cell_viability_method                            4.0  \n",
       "cell_line.cell_morphology.cell_viability_result                            4.0  \n",
       "cell_line.cell_morphology.percent_necrosis                                 4.0  \n",
       "cell_line.growth_conditions.passage_number                                 4.0  \n",
       "cell_line.growth_conditions.growth_medium                                  4.0  \n",
       "cell_line.growth_conditions.culture_environment                            4.0  \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                         4.0  \n",
       "cell_line.growth_conditions.mycoplasma_testing_...                         4.0  \n",
       "cell_line.growth_conditions.drug_treatment                                 4.0  \n",
       "cell_line.growth_conditions.feeder_layer_type                              4.0  \n",
       "cell_line.confluency                                                       4.0  \n",
       "cell_line.cell_type.text                                                   4.0  \n",
       "cell_line.cell_type.ontology                                               4.0  \n",
       "cell_line.cell_type.ontology_label                                         4.0  \n",
       "cell_line.karyotype                                                        4.0  \n",
       "cell_line.tissue.text                                                      4.0  \n",
       "cell_line.tissue.ontology                                                  4.0  \n",
       "cell_line.tissue.ontology_label                                            4.0  \n",
       "cell_line.date_established                                                 4.0  \n",
       "cell_line.disease.text                                                     4.0  \n",
       "cell_line.disease.ontology                                                 4.0  \n",
       "cell_line.disease.ontology_label                                           4.0  \n",
       "cell_line.genus_species.text                                               4.0  \n",
       "cell_line.genus_species.ontology                                           4.0  \n",
       "cell_line.genus_species.ontology_label                                     4.0  \n",
       "cell_line.publication.authors                                              4.0  \n",
       "cell_line.publication.publication_title                                    4.0  \n",
       "cell_line.publication.doi                                                  4.0  \n",
       "cell_line.publication.pmid                                                 4.0  \n",
       "cell_line.publication.publication_url                                      4.0  \n",
       "\n",
       "[590 rows x 4 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Alternatively skip above steps and import from previous 'project_counts.csv'\n",
    "import pandas as pd\n",
    "count_schema = pd.read_csv('project_counts.csv', index_col = 0)\n",
    "count_schema"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.\n",
    "\n",
    "Lets visualise the results with some interactive plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "    }\n",
       "    finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.info(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(js_urls, callback) {\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = js_urls.length;\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var s = document.createElement('script');\n",
       "      s.src = url;\n",
       "      s.async = false;\n",
       "      s.onreadystatechange = s.onload = function() {\n",
       "        root._bokeh_is_loading--;\n",
       "        if (root._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "          run_callbacks()\n",
       "        }\n",
       "      };\n",
       "      s.onerror = function() {\n",
       "        console.warn(\"failed to load library \" + url);\n",
       "      };\n",
       "      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1001\");\n",
       "  if (element == null) {\n",
       "    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.0.4.min.js\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(js_urls, function() {\n",
       "      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n    }\n    finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.info(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(js_urls, callback) {\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = js_urls.length;\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var s = document.createElement('script');\n      s.src = url;\n      s.async = false;\n      s.onreadystatechange = s.onload = function() {\n        root._bokeh_is_loading--;\n        if (root._bokeh_is_loading === 0) {\n          console.log(\"Bokeh: all BokehJS libraries loaded\");\n          run_callbacks()\n        }\n      };\n      s.onerror = function() {\n        console.warn(\"failed to load library \" + url);\n      };\n      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.getElementsByTagName(\"head\")[0].appendChild(s);\n    }\n  };var element = document.getElementById(\"1001\");\n  if (element == null) {\n    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n    return false;\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.0.4.min.js\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n    }\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(js_urls, function() {\n      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"8e05eba9-876c-4bd5-bca5-92907abb0e0a\" data-root-id=\"1002\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"14910cd9-5e57-4acc-89bf-315a081e20e7\":{\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"1011\",\"type\":\"CategoricalAxis\"}],\"left\":[{\"id\":\"1015\",\"type\":\"LinearAxis\"}],\"plot_width\":800,\"renderers\":[{\"id\":\"1011\",\"type\":\"CategoricalAxis\"},{\"id\":\"1014\",\"type\":\"Grid\"},{\"id\":\"1015\",\"type\":\"LinearAxis\"},{\"id\":\"1019\",\"type\":\"Grid\"},{\"id\":\"1028\",\"type\":\"BoxAnnotation\"},{\"id\":\"1038\",\"type\":\"GlyphRenderer\"}],\"title\":{\"id\":\"1040\",\"type\":\"Title\"},\"toolbar\":{\"id\":\"1026\",\"type\":\"Toolbar\"},\"x_range\":{\"id\":\"1003\",\"type\":\"FactorRange\"},\"x_scale\":{\"id\":\"1007\",\"type\":\"CategoricalScale\"},\"y_range\":{\"id\":\"1005\",\"type\":\"DataRange1d\"},\"y_scale\":{\"id\":\"1009\",\"type\":\"LinearScale\"}},\"id\":\"1002\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{},\"id\":\"1023\",\"type\":\"SaveTool\"},{\"attributes\":{\"plot\":null,\"text\":\"Schema usage in datasets selected\"},\"id\":\"1040\",\"type\":\"Title\"},{\"attributes\":{},\"id\":\"1024\",\"type\":\"ResetTool\"},{\"attributes\":{},\"id\":\"1025\",\"type\":\"HelpTool\"},{\"attributes\":{\"callback\":null,\"factors\":[\"sequencing_protocol\",\"project\",\"cell_suspension\",\"donor_organism\",\"specimen_from_organism\",\"sequence_file\",\"process\",\"library_preparation_protocol\",\"dissociation_protocol\",\"enrichment_protocol\",\"analysis_file\",\"analysis_process\",\"analysis_protocol\",\"collection_protocol\",\"supplementary_file\",\"cell_line\",\"differentiation_protocol\",\"ipsc_induction_protocol\",\"organoid\",\"protocol\",\"aggregate_generation_protocol\",\"reference_file\",\"protocol_file\",\"image_file\",\"imaged_specimen\",\"imaging_preparation_protocol\",\"imaging_protocol\"]},\"id\":\"1003\",\"type\":\"FactorRange\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"1020\",\"type\":\"PanTool\"},{\"id\":\"1021\",\"type\":\"WheelZoomTool\"},{\"id\":\"1022\",\"type\":\"BoxZoomTool\"},{\"id\":\"1023\",\"type\":\"SaveTool\"},{\"id\":\"1024\",\"type\":\"ResetTool\"},{\"id\":\"1025\",\"type\":\"HelpTool\"}]},\"id\":\"1026\",\"type\":\"Toolbar\"},{\"attributes\":{\"callback\":null},\"id\":\"1005\",\"type\":\"DataRange1d\"},{\"attributes\":{},\"id\":\"1007\",\"type\":\"CategoricalScale\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":{\"value\":0.5},\"fill_color\":{\"value\":\"lightgrey\"},\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":{\"value\":1.0},\"line_color\":{\"value\":\"black\"},\"line_dash\":[4,4],\"line_width\":{\"value\":2},\"plot\":null,\"render_mode\":\"css\",\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"1028\",\"type\":\"BoxAnnotation\"},{\"attributes\":{},\"id\":\"1009\",\"type\":\"LinearScale\"},{\"attributes\":{\"source\":{\"id\":\"1035\",\"type\":\"ColumnDataSource\"}},\"id\":\"1039\",\"type\":\"CDSView\"},{\"attributes\":{\"axis_label\":\"Schema\",\"axis_label_text_font_size\":{\"value\":\"12pt\"},\"formatter\":{\"id\":\"1045\",\"type\":\"CategoricalTickFormatter\"},\"major_label_orientation\":1.208304866765305,\"major_label_text_font_size\":{\"value\":\"11pt\"},\"plot\":{\"id\":\"1002\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"1012\",\"type\":\"CategoricalTicker\"}},\"id\":\"1011\",\"type\":\"CategoricalAxis\"},{\"attributes\":{},\"id\":\"1012\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"callback\":null,\"data\":{\"top\":[15.0,15.0,15.0,15.0,15.0,15.0,15.0,15.0,13.0,8.0,7.0,7.0,7.0,5.0,4.0,4.0,3.0,2.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"x\":[\"sequencing_protocol\",\"project\",\"cell_suspension\",\"donor_organism\",\"specimen_from_organism\",\"sequence_file\",\"process\",\"library_preparation_protocol\",\"dissociation_protocol\",\"enrichment_protocol\",\"analysis_file\",\"analysis_process\",\"analysis_protocol\",\"collection_protocol\",\"supplementary_file\",\"cell_line\",\"differentiation_protocol\",\"ipsc_induction_protocol\",\"organoid\",\"protocol\",\"aggregate_generation_protocol\",\"reference_file\",\"protocol_file\",\"image_file\",\"imaged_specimen\",\"imaging_preparation_protocol\",\"imaging_protocol\"]},\"selected\":{\"id\":\"1047\",\"type\":\"Selection\"},\"selection_policy\":{\"id\":\"1048\",\"type\":\"UnionRenderers\"}},\"id\":\"1035\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"1043\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"plot\":{\"id\":\"1002\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"1012\",\"type\":\"CategoricalTicker\"}},\"id\":\"1014\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"1045\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{\"axis_label\":\"No. of projects using schema\",\"axis_label_text_font_size\":{\"value\":\"12pt\"},\"formatter\":{\"id\":\"1043\",\"type\":\"BasicTickFormatter\"},\"plot\":{\"id\":\"1002\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"1016\",\"type\":\"BasicTicker\"}},\"id\":\"1015\",\"type\":\"LinearAxis\"},{\"attributes\":{},\"id\":\"1047\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"1016\",\"type\":\"BasicTicker\"},{\"attributes\":{},\"id\":\"1048\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"dimension\":1,\"plot\":{\"id\":\"1002\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"1016\",\"type\":\"BasicTicker\"}},\"id\":\"1019\",\"type\":\"Grid\"},{\"attributes\":{\"data_source\":{\"id\":\"1035\",\"type\":\"ColumnDataSource\"},\"glyph\":{\"id\":\"1036\",\"type\":\"VBar\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1037\",\"type\":\"VBar\"},\"selection_glyph\":null,\"view\":{\"id\":\"1039\",\"type\":\"CDSView\"}},\"id\":\"1038\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"top\":{\"field\":\"top\"},\"width\":{\"value\":0.9},\"x\":{\"field\":\"x\"}},\"id\":\"1037\",\"type\":\"VBar\"},{\"attributes\":{\"fill_color\":{\"value\":\"green\"},\"line_color\":{\"value\":\"green\"},\"top\":{\"field\":\"top\"},\"width\":{\"value\":0.9},\"x\":{\"field\":\"x\"}},\"id\":\"1036\",\"type\":\"VBar\"},{\"attributes\":{},\"id\":\"1020\",\"type\":\"PanTool\"},{\"attributes\":{},\"id\":\"1021\",\"type\":\"WheelZoomTool\"},{\"attributes\":{\"overlay\":{\"id\":\"1028\",\"type\":\"BoxAnnotation\"}},\"id\":\"1022\",\"type\":\"BoxZoomTool\"}],\"root_ids\":[\"1002\"]},\"title\":\"Bokeh Application\",\"version\":\"1.0.4\"}};\n",
       "  var render_items = [{\"docid\":\"14910cd9-5e57-4acc-89bf-315a081e20e7\",\"roots\":{\"1002\":\"8e05eba9-876c-4bd5-bca5-92907abb0e0a\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        embed_document(root);\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "      attempts++;\n",
       "      if (attempts > 100) {\n",
       "        console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "1002"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# format dataframe\n",
    "schema_usage = (\n",
    "                count_schema[['Schema Name', 'Schema usage (no. project)']]\n",
    "                .drop_duplicates()\n",
    "                .sort_values(by='Schema usage (no. project)', ascending = False)\n",
    "                .set_index(['Schema Name'])\n",
    "               )\n",
    "\n",
    "# plotting\n",
    "\n",
    "from bokeh.io import show, output_notebook, reset_output\n",
    "from bokeh.plotting import figure\n",
    "from bokeh.models import ColumnDataSource\n",
    "import math\n",
    "\n",
    "\n",
    "output_notebook()\n",
    "\n",
    "# source = ColumnDataSource(schema_usage)\n",
    "x = schema_usage.index.tolist()\n",
    "y = schema_usage['Schema usage (no. project)'].tolist()\n",
    "\n",
    "\n",
    "p = figure(plot_width=800, plot_height=600, x_range=x)\n",
    "\n",
    "p.vbar(x=x, top=y, width=0.9, color='green')\n",
    "\n",
    "p.title.text = 'Schema usage in datasets selected'\n",
    "p.xaxis.axis_label = 'Schema'\n",
    "p.yaxis.axis_label = 'No. of projects using schema'\n",
    "p.xaxis.major_label_orientation = math.pi/2.6\n",
    "p.yaxis.axis_label_text_font_size = \"12pt\"\n",
    "p.xaxis.axis_label_text_font_size = \"12pt\"\n",
    "p.xaxis.major_label_text_font_size = \"11pt\"\n",
    "\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1104\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1104\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "    }\n",
       "    finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.info(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(js_urls, callback) {\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = js_urls.length;\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var s = document.createElement('script');\n",
       "      s.src = url;\n",
       "      s.async = false;\n",
       "      s.onreadystatechange = s.onload = function() {\n",
       "        root._bokeh_is_loading--;\n",
       "        if (root._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "          run_callbacks()\n",
       "        }\n",
       "      };\n",
       "      s.onerror = function() {\n",
       "        console.warn(\"failed to load library \" + url);\n",
       "      };\n",
       "      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1104\");\n",
       "  if (element == null) {\n",
       "    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '1104' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.0.4.min.js\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1104\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(js_urls, function() {\n",
       "      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1104\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n    }\n    finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.info(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(js_urls, callback) {\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = js_urls.length;\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var s = document.createElement('script');\n      s.src = url;\n      s.async = false;\n      s.onreadystatechange = s.onload = function() {\n        root._bokeh_is_loading--;\n        if (root._bokeh_is_loading === 0) {\n          console.log(\"Bokeh: all BokehJS libraries loaded\");\n          run_callbacks()\n        }\n      };\n      s.onerror = function() {\n        console.warn(\"failed to load library \" + url);\n      };\n      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.getElementsByTagName(\"head\")[0].appendChild(s);\n    }\n  };var element = document.getElementById(\"1104\");\n  if (element == null) {\n    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '1104' but no matching script tag was found. \")\n    return false;\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.0.4.min.js\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.4.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.4.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.4.min.css\");\n    }\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1104\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(js_urls, function() {\n      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Select schema from list.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dcadb4e5f0af4110a04c892d9e368b84",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='schema', options=('ipsc_induction_protocol', 'cell_line', 'referen"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<function __main__.update(schema)>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "import bokeh.plotting as bp\n",
    "from bokeh.io import show\n",
    "from bokeh.models.ranges import FactorRange\n",
    "from bokeh.models.glyphs import VBar\n",
    "from ipywidgets import interact\n",
    "output_notebook()\n",
    "\n",
    "schema_names = list(set(count_schema['Schema Name'].tolist()))\n",
    "f = schema_names[3] # first iteration\n",
    "\n",
    "\n",
    "def get_data(f):\n",
    "    \n",
    "    data = (count_schema.loc[count_schema['Schema Name'] == f]\n",
    "            .reset_index()\n",
    "            .sort_values(by='Attribute usage (no. projects)', ascending = False))\n",
    "    \n",
    "    def shorter_name(attribute_name): # add column shorter attribute names missing top schema name\n",
    "        levels = attribute_name.split('.')\n",
    "        return '.'.join(levels[1:])\n",
    "    data['x_label'] = data['index'].apply(shorter_name)\n",
    "    return data\n",
    "\n",
    "\n",
    "def plotter(data, f):\n",
    "\n",
    "    source = ColumnDataSource(data)\n",
    "    \n",
    "    \n",
    "    p = figure(plot_width=800,\n",
    "               plot_height=600,\n",
    "               x_range=FactorRange(factors=list(data['x_label'])),\n",
    "               background_fill_color='#efefef')\n",
    "    \n",
    "    r = p.vbar(x=\"x_label\", top=\"Attribute usage (no. projects)\", source=source, bottom=0, width=0.5, fill_color=\"#b3de69\")\n",
    "    \n",
    "\n",
    "    p.title.text = 'Schema usage in datasets selected'\n",
    "    p.yaxis.axis_label = 'No. of projects using attribute'\n",
    "    p.xaxis.major_label_orientation = math.pi/2.6\n",
    "    p.xaxis.axis_label = f + ' attributes'\n",
    "    p.yaxis.axis_label_text_font_size = \"12pt\"\n",
    "    p.xaxis.axis_label_text_font_size = \"12pt\"\n",
    "    p.xaxis.major_label_text_font_size = \"11pt\"\n",
    "    return (p)\n",
    "\n",
    "def update(schema):\n",
    "    data = get_data(schema)\n",
    "    plotting = plotter(data, schema)\n",
    "    p = plotting\n",
    "    handle = show(p, notebook_handle=True)\n",
    "\n",
    "print('Select schema from list.')\n",
    "interact(update, schema=schema_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
