{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metadata Vignette\n",
    "\n",
    "## Attribute differences between a submission and latest schema.\n",
    "\n",
    "### By Matthew Green\n",
    "\n",
    "\n",
    "As metadata evolves projects in the datastore can get out of date with current schema. This vignette script takes a project uuid and looks for differences between it's attributes as submirtted versus latest schema. It also suggests possible swaps by simple lexical similarity lookup. This can be a useful tool for wranglers to use when they are updating a dataset.\n",
    "\n",
    "\n",
    "**Pseudocode**\n",
    "\n",
    "1. User provides project uuid\n",
    "2. Get latest attributes\n",
    "3. Get attributes from provided project. (bundlewise search)\n",
    "4. Diff\n",
    "5. Suggest lexically similar replacement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Original cBeta Submissions\n",
    "\n",
    "treutlein: e8642221-4c2c-4fd7-b926-a68bce363c88\n",
    "meyer: 5dfe932f-159d-4cab-8039-d32f22ffbbc2\n",
    "peer: 29f53b7e-071b-44b5-998a-0ae70d0229a4\n",
    "neuron_diff: f8880be0-210c-4aa3-9348-f5a423e07421\n",
    "fetal-maternal-interface: aabbec1a-1215-43e1-8e42-6489af25c12c\n",
    "Regev-ICA: 179bf9e6-5b33-4c5b-ae26-96c7270976b8 (IndexError: list index out of range)\n",
    "EMTAB5061: 1a0f98b8-746a-489d-8af9-d5c657482aab\n",
    "Teichmann-mouse-melanoma: f396fa53-2a2d-4b8a-ad18-03bf4bd46833\n",
    "EGEOD106540: 0ec2b05f-ddbe-4e5a-b30f-e81f4b1e330c\n",
    "tabulamuris: 32eb86db-6842-480f-a49a-a2b0161ed35a (NO BUNDLES)\n",
    "ido_amit: 0c7bbbce-3c70-4d6b-a443-1b92c1f205c8 (HTTP 502)\n",
    "humphreys: 34ec62a2-9643-430d-b41a-1e342bd615fc\n",
    "basu: c765e3f9-7cfc-4501-8832-79e5f7abd321\n",
    "10x-mouse-brain: ff481f29-3d0b-4533-9de2-a760c61c162d (IndexError: list index out of range)\n",
    "rsatija: 5f256182-5dfc-4070-8404-f6fa71d37c73\n",
    "\n",
    "Provide a uuid to analyse:\n",
    "'''\n",
    "#########################\n",
    "# EDIT PROJECT UUID HERE\n",
    "#########################\n",
    "\n",
    "query_uuid = '5f256182-5dfc-4070-8404-f6fa71d37c73'\n",
    "\n",
    "#####################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "590 out of 590 attributes are unique\n",
      "['sequencing_protocol.protocol_core.protocol_id', 'sequencing_protocol.protocol_core.protocol_name', 'sequencing_protocol.protocol_core.protocol_description', 'sequencing_protocol.protocol_core.publication_doi', 'sequencing_protocol.protocol_core.protocols_io_doi', 'sequencing_protocol.protocol_core.document', 'sequencing_protocol.instrument_manufacturer_model.text', 'sequencing_protocol.instrument_manufacturer_model.ontology', 'sequencing_protocol.instrument_manufacturer_model.ontology_label', 'sequencing_protocol.local_machine_name', 'sequencing_protocol.paired_end', 'sequencing_protocol.method.text', 'sequencing_protocol.method.ontology', 'sequencing_protocol.method.ontology_label', 'sequencing_protocol.10x.fastq_method', 'sequencing_protocol.10x.fastq_method_version', 'sequencing_protocol.10x.pooled_channels', 'sequencing_protocol.10x.drop_uniformity', 'library_preparation_protocol.protocol_core.protocol_id', 'library_preparation_protocol.protocol_core.protocol_name', 'library_preparation_protocol.protocol_core.protocol_description', 'library_preparation_protocol.protocol_core.publication_doi', 'library_preparation_protocol.protocol_core.protocols_io_doi', 'library_preparation_protocol.protocol_core.document', 'library_preparation_protocol.cell_barcode.barcode_read', 'library_preparation_protocol.cell_barcode.barcode_offset', 'library_preparation_protocol.cell_barcode.barcode_length', 'library_preparation_protocol.cell_barcode.white_list_file', 'library_preparation_protocol.input_nucleic_acid_molecule.text', 'library_preparation_protocol.input_nucleic_acid_molecule.ontology', 'library_preparation_protocol.input_nucleic_acid_molecule.ontology_label', 'library_preparation_protocol.nucleic_acid_source', 'library_preparation_protocol.library_construction_method.text', 'library_preparation_protocol.library_construction_method.ontology', 'library_preparation_protocol.library_construction_method.ontology_label', 'library_preparation_protocol.library_construction_kit.retail_name', 'library_preparation_protocol.library_construction_kit.catalog_number', 'library_preparation_protocol.library_construction_kit.manufacturer', 'library_preparation_protocol.library_construction_kit.lot_number', 'library_preparation_protocol.library_construction_kit.expiry_date', 'library_preparation_protocol.library_construction_kit.kit_titer', 'library_preparation_protocol.nucleic_acid_conversion_kit.retail_name', 'library_preparation_protocol.nucleic_acid_conversion_kit.catalog_number', 'library_preparation_protocol.nucleic_acid_conversion_kit.manufacturer', 'library_preparation_protocol.nucleic_acid_conversion_kit.lot_number', 'library_preparation_protocol.nucleic_acid_conversion_kit.expiry_date', 'library_preparation_protocol.nucleic_acid_conversion_kit.kit_titer', 'library_preparation_protocol.end_bias', 'library_preparation_protocol.primer', 'library_preparation_protocol.strand', 'library_preparation_protocol.spike_in_kit.retail_name', 'library_preparation_protocol.spike_in_kit.catalog_number', 'library_preparation_protocol.spike_in_kit.manufacturer', 'library_preparation_protocol.spike_in_kit.lot_number', 'library_preparation_protocol.spike_in_kit.expiry_date', 'library_preparation_protocol.spike_in_kit.kit_titer', 'library_preparation_protocol.spike_in_dilution', 'library_preparation_protocol.umi_barcode.barcode_read', 'library_preparation_protocol.umi_barcode.barcode_offset', 'library_preparation_protocol.umi_barcode.barcode_length', 'library_preparation_protocol.umi_barcode.white_list_file', 'library_preparation_protocol.library_preamplification_method.text', 'library_preparation_protocol.library_preamplification_method.ontology', 'library_preparation_protocol.library_preamplification_method.ontology_label', 'library_preparation_protocol.cdna_library_amplification_method.text', 'library_preparation_protocol.cdna_library_amplification_method.ontology', 'library_preparation_protocol.cdna_library_amplification_method.ontology_label', 'library_preparation_protocol.nominal_length', 'library_preparation_protocol.nominal_sdev', 'imaging_protocol.protocol_core.protocol_id', 'imaging_protocol.protocol_core.protocol_name', 'imaging_protocol.protocol_core.protocol_description', 'imaging_protocol.protocol_core.publication_doi', 'imaging_protocol.protocol_core.protocols_io_doi', 'imaging_protocol.protocol_core.document', 'imaging_protocol.microscope_setup_description', 'imaging_protocol.microscopy_technique.text', 'imaging_protocol.microscopy_technique.ontology', 'imaging_protocol.microscopy_technique.ontology_label', 'imaging_protocol.magnification', 'imaging_protocol.numerical_aperture', 'imaging_protocol.immersion_medium_type', 'imaging_protocol.immersion_medium_refractive_index', 'imaging_protocol.pixel_size', 'imaging_protocol.number_of_tiles', 'imaging_protocol.tile_size_y', 'imaging_protocol.tile_size_x', 'imaging_protocol.z_stack_step_size', 'imaging_protocol.number_of_z_steps', 'imaging_protocol.overlapping_tiles', 'imaging_protocol.channel.channel_id', 'imaging_protocol.channel.excitation_wavelength', 'imaging_protocol.channel.filter_range', 'imaging_protocol.channel.multiplexed', 'imaging_protocol.channel.target_fluorophore', 'imaging_protocol.channel.exposure_time', 'imaging_protocol.target.molecule_name', 'imaging_protocol.target.molecule_id', 'imaging_protocol.target.subcellular_structure.text', 'imaging_protocol.target.subcellular_structure.ontology', 'imaging_protocol.target.subcellular_structure.ontology_label', 'imaging_protocol.target.reagent_name', 'imaging_protocol.target.purchased_reagent_details.retail_name', 'imaging_protocol.target.purchased_reagent_details.catalog_number', 'imaging_protocol.target.purchased_reagent_details.manufacturer', 'imaging_protocol.target.purchased_reagent_details.lot_number', 'imaging_protocol.target.purchased_reagent_details.expiry_date', 'imaging_protocol.target.purchased_reagent_details.kit_titer', 'imaging_protocol.target.probe_sequence', 'imaging_protocol.target.fluorophore', 'imaging_protocol.target.assay_type.text', 'imaging_protocol.target.assay_type.ontology', 'imaging_protocol.target.assay_type.ontology_label', 'imaging_protocol.target.multiplexed', 'imaging_protocol.target.channel_id', 'imaging_preparation_protocol.protocol_core.protocol_id', 'imaging_preparation_protocol.protocol_core.protocol_name', 'imaging_preparation_protocol.protocol_core.protocol_description', 'imaging_preparation_protocol.protocol_core.publication_doi', 'imaging_preparation_protocol.protocol_core.protocols_io_doi', 'imaging_preparation_protocol.protocol_core.document', 'imaging_preparation_protocol.fresh_slicing_method', 'imaging_preparation_protocol.imaged_slice_thickness', 'imaging_preparation_protocol.final_slicing_method', 'imaging_preparation_protocol.post_resection_interval', 'imaging_preparation_protocol.post_resection_interval_unit.text', 'imaging_preparation_protocol.post_resection_interval_unit.ontology', 'imaging_preparation_protocol.post_resection_interval_unit.ontology_label', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_method', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time_unit.text', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time_unit.ontology', 'imaging_preparation_protocol.pre_final_slice_preservation_method.storage_time_unit.ontology_label', 'imaging_preparation_protocol.pre_final_slice_preservation_method.preservation_method', 'imaging_preparation_protocol.post_final_slicing_interval', 'imaging_preparation_protocol.post_final_slicing_interval_unit.text', 'imaging_preparation_protocol.post_final_slicing_interval_unit.ontology', 'imaging_preparation_protocol.post_final_slicing_interval_unit.ontology_label', 'imaging_preparation_protocol.fiducial_marker', 'imaging_preparation_protocol.expansion_factor', 'ipsc_induction_protocol.protocol_core.protocol_id', 'ipsc_induction_protocol.protocol_core.protocol_name', 'ipsc_induction_protocol.protocol_core.protocol_description', 'ipsc_induction_protocol.protocol_core.publication_doi', 'ipsc_induction_protocol.protocol_core.protocols_io_doi', 'ipsc_induction_protocol.protocol_core.document', 'ipsc_induction_protocol.ipsc_induction_method', 'ipsc_induction_protocol.ipsc_induction_factors', 'ipsc_induction_protocol.ipsc_induction_kit.retail_name', 'ipsc_induction_protocol.ipsc_induction_kit.catalog_number', 'ipsc_induction_protocol.ipsc_induction_kit.manufacturer', 'ipsc_induction_protocol.ipsc_induction_kit.lot_number', 'ipsc_induction_protocol.ipsc_induction_kit.expiry_date', 'ipsc_induction_protocol.ipsc_induction_kit.kit_titer', 'ipsc_induction_protocol.pluripotency_test', 'ipsc_induction_protocol.percent_pluripotency', 'ipsc_induction_protocol.pluripotency_vector_removed', 'ipsc_induction_protocol.ipsc_induction_produced_in_house', 'ipsc_induction_protocol.protocol_reagents.retail_name', 'ipsc_induction_protocol.protocol_reagents.catalog_number', 'ipsc_induction_protocol.protocol_reagents.manufacturer', 'ipsc_induction_protocol.protocol_reagents.lot_number', 'ipsc_induction_protocol.protocol_reagents.expiry_date', 'ipsc_induction_protocol.protocol_reagents.kit_titer', 'enrichment_protocol.protocol_core.protocol_id', 'enrichment_protocol.protocol_core.protocol_name', 'enrichment_protocol.protocol_core.protocol_description', 'enrichment_protocol.protocol_core.publication_doi', 'enrichment_protocol.protocol_core.protocols_io_doi', 'enrichment_protocol.protocol_core.document', 'enrichment_protocol.enrichment_method.text', 'enrichment_protocol.enrichment_method.ontology', 'enrichment_protocol.enrichment_method.ontology_label', 'enrichment_protocol.markers', 'enrichment_protocol.min_size_selected', 'enrichment_protocol.max_size_selected', 'dissociation_protocol.protocol_core.protocol_id', 'dissociation_protocol.protocol_core.protocol_name', 'dissociation_protocol.protocol_core.protocol_description', 'dissociation_protocol.protocol_core.publication_doi', 'dissociation_protocol.protocol_core.protocols_io_doi', 'dissociation_protocol.protocol_core.document', 'dissociation_protocol.method.text', 'dissociation_protocol.method.ontology', 'dissociation_protocol.method.ontology_label', 'dissociation_protocol.reagents.retail_name', 'dissociation_protocol.reagents.catalog_number', 'dissociation_protocol.reagents.manufacturer', 'dissociation_protocol.reagents.lot_number', 'dissociation_protocol.reagents.expiry_date', 'dissociation_protocol.reagents.kit_titer', 'differentiation_protocol.protocol_core.protocol_id', 'differentiation_protocol.protocol_core.protocol_name', 'differentiation_protocol.protocol_core.protocol_description', 'differentiation_protocol.protocol_core.publication_doi', 'differentiation_protocol.protocol_core.protocols_io_doi', 'differentiation_protocol.protocol_core.document', 'differentiation_protocol.differentiation_method', 'differentiation_protocol.media', 'differentiation_protocol.small_molecules', 'differentiation_protocol.target_cell_yield', 'differentiation_protocol.reagents.retail_name', 'differentiation_protocol.reagents.catalog_number', 'differentiation_protocol.reagents.manufacturer', 'differentiation_protocol.reagents.lot_number', 'differentiation_protocol.reagents.expiry_date', 'differentiation_protocol.reagents.kit_titer', 'differentiation_protocol.target_pathway', 'differentiation_protocol.validation_method', 'differentiation_protocol.validation_result', 'collection_protocol.protocol_core.protocol_id', 'collection_protocol.protocol_core.protocol_name', 'collection_protocol.protocol_core.protocol_description', 'collection_protocol.protocol_core.publication_doi', 'collection_protocol.protocol_core.protocols_io_doi', 'collection_protocol.protocol_core.document', 'collection_protocol.method.text', 'collection_protocol.method.ontology', 'collection_protocol.method.ontology_label', 'collection_protocol.reagents.retail_name', 'collection_protocol.reagents.catalog_number', 'collection_protocol.reagents.manufacturer', 'collection_protocol.reagents.lot_number', 'collection_protocol.reagents.expiry_date', 'collection_protocol.reagents.kit_titer', 'aggregate_generation_protocol.protocol_core.protocol_id', 'aggregate_generation_protocol.protocol_core.protocol_name', 'aggregate_generation_protocol.protocol_core.protocol_description', 'aggregate_generation_protocol.protocol_core.publication_doi', 'aggregate_generation_protocol.protocol_core.protocols_io_doi', 'aggregate_generation_protocol.protocol_core.document', 'aggregate_generation_protocol.formation_method', 'aggregate_generation_protocol.cell_uniformity', 'analysis_protocol.protocol_core.protocol_id', 'analysis_protocol.protocol_core.protocol_name', 'analysis_protocol.protocol_core.protocol_description', 'analysis_protocol.protocol_core.publication_doi', 'analysis_protocol.protocol_core.protocols_io_doi', 'analysis_protocol.protocol_core.document', 'analysis_protocol.protocol_type.text', 'analysis_protocol.protocol_type.ontology', 'analysis_protocol.protocol_type.ontology_label', 'analysis_protocol.computational_method', 'protocol.protocol_core.protocol_id', 'protocol.protocol_core.protocol_name', 'protocol.protocol_core.protocol_description', 'protocol.protocol_core.publication_doi', 'protocol.protocol_core.protocols_io_doi', 'protocol.protocol_core.document', 'protocol.protocol_type.text', 'protocol.protocol_type.ontology', 'protocol.protocol_type.ontology_label', 'project.project_core.project_short_name', 'project.project_core.project_title', 'project.project_core.project_description', 'project.contributors.contact_name', 'project.contributors.email', 'project.contributors.phone', 'project.contributors.institution', 'project.contributors.laboratory', 'project.contributors.address', 'project.contributors.country', 'project.contributors.corresponding_contributor', 'project.contributors.project_role', 'project.contributors.orcid_id', 'project.supplementary_links', 'project.publications.authors', 'project.publications.publication_title', 'project.publications.doi', 'project.publications.pmid', 'project.publications.publication_url', 'project.insdc_project_accessions', 'project.geo_series_accessions', 'project.array_express_accessions', 'project.insdc_study_accessions', 'project.funders.grant_title', 'project.funders.grant_id', 'project.funders.organization', 'analysis_process.process_core.process_id', 'analysis_process.process_core.process_name', 'analysis_process.process_core.process_description', 'analysis_process.process_core.process_location', 'analysis_process.process_core.operator', 'analysis_process.process_type.text', 'analysis_process.process_type.ontology', 'analysis_process.process_type.ontology_label', 'analysis_process.inputs.checksum', 'analysis_process.inputs.parameter_name', 'analysis_process.inputs.parameter_value', 'analysis_process.tasks.disk_size', 'analysis_process.tasks.task_name', 'analysis_process.tasks.zone', 'analysis_process.tasks.log_err', 'analysis_process.tasks.start_time', 'analysis_process.tasks.cpus', 'analysis_process.tasks.log_out', 'analysis_process.tasks.stop_time', 'analysis_process.tasks.memory', 'analysis_process.tasks.docker_image', 'analysis_process.input_bundles', 'analysis_process.timestamp_start_utc', 'analysis_process.timestamp_stop_utc', 'analysis_process.analysis_run_type', 'analysis_process.reference_bundle', 'process.process_core.process_id', 'process.process_core.process_name', 'process.process_core.process_description', 'process.process_core.process_location', 'process.process_core.operator', 'process.length_of_time', 'process.length_of_time_unit.text', 'process.length_of_time_unit.ontology', 'process.length_of_time_unit.ontology_label', 'process.process_type.text', 'process.process_type.ontology', 'process.process_type.ontology_label', 'process.deviation_from_protocol', 'process.insdc_experiment.insdc_experiment_accession', 'supplementary_file.file_core.file_name', 'supplementary_file.file_core.file_format', 'supplementary_file.file_core.checksum', 'supplementary_file.file_description', 'sequence_file.file_core.file_name', 'sequence_file.file_core.file_format', 'sequence_file.file_core.checksum', 'sequence_file.read_index', 'sequence_file.lane_index', 'sequence_file.read_length', 'sequence_file.insdc_run_accessions', 'sequence_file.library_prep_id', 'reference_file.file_core.file_name', 'reference_file.file_core.file_format', 'reference_file.file_core.checksum', 'reference_file.ncbi_taxon_id', 'reference_file.genus_species.text', 'reference_file.genus_species.ontology', 'reference_file.genus_species.ontology_label', 'reference_file.reference_type', 'reference_file.assembly_type', 'reference_file.reference_version', 'protocol_file.file_core.file_name', 'protocol_file.file_core.file_format', 'protocol_file.file_core.checksum', 'image_file.file_core.file_name', 'image_file.file_core.file_format', 'image_file.file_core.checksum', 'analysis_file.file_core.file_name', 'analysis_file.file_core.file_format', 'analysis_file.file_core.checksum', 'specimen_from_organism.biomaterial_core.biomaterial_id', 'specimen_from_organism.biomaterial_core.biomaterial_name', 'specimen_from_organism.biomaterial_core.biomaterial_description', 'specimen_from_organism.biomaterial_core.ncbi_taxon_id', 'specimen_from_organism.biomaterial_core.genotype', 'specimen_from_organism.biomaterial_core.supplementary_files', 'specimen_from_organism.biomaterial_core.biosd_biomaterial', 'specimen_from_organism.biomaterial_core.insdc_biomaterial', 'specimen_from_organism.genus_species.text', 'specimen_from_organism.genus_species.ontology', 'specimen_from_organism.genus_species.ontology_label', 'specimen_from_organism.organ.text', 'specimen_from_organism.organ.ontology', 'specimen_from_organism.organ.ontology_label', 'specimen_from_organism.organ_parts.text', 'specimen_from_organism.organ_parts.ontology', 'specimen_from_organism.organ_parts.ontology_label', 'specimen_from_organism.diseases.text', 'specimen_from_organism.diseases.ontology', 'specimen_from_organism.diseases.ontology_label', 'specimen_from_organism.state_of_specimen.autolysis_score', 'specimen_from_organism.state_of_specimen.gross_description', 'specimen_from_organism.state_of_specimen.gross_images', 'specimen_from_organism.state_of_specimen.ischemic_temperature', 'specimen_from_organism.state_of_specimen.ischemic_time', 'specimen_from_organism.state_of_specimen.microscopic_description', 'specimen_from_organism.state_of_specimen.microscopic_images', 'specimen_from_organism.state_of_specimen.postmortem_interval', 'specimen_from_organism.preservation_storage.storage_method', 'specimen_from_organism.preservation_storage.storage_time', 'specimen_from_organism.preservation_storage.storage_time_unit.text', 'specimen_from_organism.preservation_storage.storage_time_unit.ontology', 'specimen_from_organism.preservation_storage.storage_time_unit.ontology_label', 'specimen_from_organism.preservation_storage.preservation_method', 'specimen_from_organism.collection_time', 'specimen_from_organism.purchased_specimen.retail_name', 'specimen_from_organism.purchased_specimen.catalog_number', 'specimen_from_organism.purchased_specimen.manufacturer', 'specimen_from_organism.purchased_specimen.lot_number', 'specimen_from_organism.purchased_specimen.expiry_date', 'specimen_from_organism.purchased_specimen.kit_titer', 'organoid.biomaterial_core.biomaterial_id', 'organoid.biomaterial_core.biomaterial_name', 'organoid.biomaterial_core.biomaterial_description', 'organoid.biomaterial_core.ncbi_taxon_id', 'organoid.biomaterial_core.genotype', 'organoid.biomaterial_core.supplementary_files', 'organoid.biomaterial_core.biosd_biomaterial', 'organoid.biomaterial_core.insdc_biomaterial', 'organoid.genus_species.text', 'organoid.genus_species.ontology', 'organoid.genus_species.ontology_label', 'organoid.model_organ.text', 'organoid.model_organ.ontology', 'organoid.model_organ.ontology_label', 'organoid.model_organ_part.text', 'organoid.model_organ_part.ontology', 'organoid.model_organ_part.ontology_label', 'organoid.age', 'organoid.age_unit.text', 'organoid.age_unit.ontology', 'organoid.age_unit.ontology_label', 'organoid.size', 'organoid.size_unit.text', 'organoid.size_unit.ontology', 'organoid.size_unit.ontology_label', 'organoid.morphology', 'organoid.embedded_in_matrigel', 'organoid.growth_environment', 'organoid.input_aggregate_cell_count', 'organoid.stored_oxygen_levels', 'imaged_specimen.biomaterial_core.biomaterial_id', 'imaged_specimen.biomaterial_core.biomaterial_name', 'imaged_specimen.biomaterial_core.biomaterial_description', 'imaged_specimen.biomaterial_core.ncbi_taxon_id', 'imaged_specimen.biomaterial_core.genotype', 'imaged_specimen.biomaterial_core.supplementary_files', 'imaged_specimen.biomaterial_core.biosd_biomaterial', 'imaged_specimen.biomaterial_core.insdc_biomaterial', 'imaged_specimen.overview_images', 'imaged_specimen.slice_thickness', 'imaged_specimen.internal_anatomical_structures.text', 'imaged_specimen.internal_anatomical_structures.ontology', 'imaged_specimen.internal_anatomical_structures.ontology_label', 'donor_organism.biomaterial_core.biomaterial_id', 'donor_organism.biomaterial_core.biomaterial_name', 'donor_organism.biomaterial_core.biomaterial_description', 'donor_organism.biomaterial_core.ncbi_taxon_id', 'donor_organism.biomaterial_core.genotype', 'donor_organism.biomaterial_core.supplementary_files', 'donor_organism.biomaterial_core.biosd_biomaterial', 'donor_organism.biomaterial_core.insdc_biomaterial', 'donor_organism.human_specific.body_mass_index', 'donor_organism.human_specific.ethnicity.text', 'donor_organism.human_specific.ethnicity.ontology', 'donor_organism.human_specific.ethnicity.ontology_label', 'donor_organism.mouse_specific.strain.text', 'donor_organism.mouse_specific.strain.ontology', 'donor_organism.mouse_specific.strain.ontology_label', 'donor_organism.genus_species.text', 'donor_organism.genus_species.ontology', 'donor_organism.genus_species.ontology_label', 'donor_organism.sex', 'donor_organism.is_living', 'donor_organism.organism_age', 'donor_organism.organism_age_unit.text', 'donor_organism.organism_age_unit.ontology', 'donor_organism.organism_age_unit.ontology_label', 'donor_organism.development_stage.text', 'donor_organism.development_stage.ontology', 'donor_organism.development_stage.ontology_label', 'donor_organism.diseases.text', 'donor_organism.diseases.ontology', 'donor_organism.diseases.ontology_label', 'donor_organism.death.cause_of_death', 'donor_organism.death.cold_perfused', 'donor_organism.death.days_on_ventilator', 'donor_organism.death.hardy_scale', 'donor_organism.death.time_of_death', 'donor_organism.death.organ_donation_death_type', 'donor_organism.death.normothermic_regional_perfusion', 'donor_organism.familial_relationships.parent', 'donor_organism.familial_relationships.child', 'donor_organism.familial_relationships.sibling', 'donor_organism.medical_history.alcohol_history', 'donor_organism.medical_history.medication', 'donor_organism.medical_history.smoking_history', 'donor_organism.medical_history.nutritional_state', 'donor_organism.medical_history.test_results', 'donor_organism.medical_history.treatment', 'donor_organism.gestational_age', 'donor_organism.gestational_age_unit.text', 'donor_organism.gestational_age_unit.ontology', 'donor_organism.gestational_age_unit.ontology_label', 'donor_organism.height', 'donor_organism.height_unit.text', 'donor_organism.height_unit.ontology', 'donor_organism.height_unit.ontology_label', 'donor_organism.weight', 'donor_organism.weight_unit.text', 'donor_organism.weight_unit.ontology', 'donor_organism.weight_unit.ontology_label', 'donor_organism.timecourse.value', 'donor_organism.timecourse.unit.text', 'donor_organism.timecourse.unit.ontology', 'donor_organism.timecourse.unit.ontology_label', 'donor_organism.timecourse.relevance', 'cell_suspension.biomaterial_core.biomaterial_id', 'cell_suspension.biomaterial_core.biomaterial_name', 'cell_suspension.biomaterial_core.biomaterial_description', 'cell_suspension.biomaterial_core.ncbi_taxon_id', 'cell_suspension.biomaterial_core.genotype', 'cell_suspension.biomaterial_core.supplementary_files', 'cell_suspension.biomaterial_core.biosd_biomaterial', 'cell_suspension.biomaterial_core.insdc_biomaterial', 'cell_suspension.cell_morphology.cell_morphology', 'cell_suspension.cell_morphology.cell_size', 'cell_suspension.cell_morphology.cell_size_unit.text', 'cell_suspension.cell_morphology.cell_size_unit.ontology', 'cell_suspension.cell_morphology.cell_size_unit.ontology_label', 'cell_suspension.cell_morphology.percent_cell_viability', 'cell_suspension.cell_morphology.cell_viability_method', 'cell_suspension.cell_morphology.cell_viability_result', 'cell_suspension.cell_morphology.percent_necrosis', 'cell_suspension.growth_conditions.passage_number', 'cell_suspension.growth_conditions.growth_medium', 'cell_suspension.growth_conditions.culture_environment', 'cell_suspension.growth_conditions.mycoplasma_testing_method', 'cell_suspension.growth_conditions.mycoplasma_testing_results', 'cell_suspension.growth_conditions.drug_treatment', 'cell_suspension.growth_conditions.feeder_layer_type', 'cell_suspension.genus_species.text', 'cell_suspension.genus_species.ontology', 'cell_suspension.genus_species.ontology_label', 'cell_suspension.selected_cell_type.text', 'cell_suspension.selected_cell_type.ontology', 'cell_suspension.selected_cell_type.ontology_label', 'cell_suspension.estimated_cell_count', 'cell_suspension.plate_based_sequencing.plate_label', 'cell_suspension.plate_based_sequencing.well_label', 'cell_suspension.plate_based_sequencing.well_quality', 'cell_suspension.timecourse.value', 'cell_suspension.timecourse.unit.text', 'cell_suspension.timecourse.unit.ontology', 'cell_suspension.timecourse.unit.ontology_label', 'cell_suspension.timecourse.relevance', 'cell_line.biomaterial_core.biomaterial_id', 'cell_line.biomaterial_core.biomaterial_name', 'cell_line.biomaterial_core.biomaterial_description', 'cell_line.biomaterial_core.ncbi_taxon_id', 'cell_line.biomaterial_core.genotype', 'cell_line.biomaterial_core.supplementary_files', 'cell_line.biomaterial_core.biosd_biomaterial', 'cell_line.biomaterial_core.insdc_biomaterial', 'cell_line.supplier', 'cell_line.catalog_number', 'cell_line.lot_number', 'cell_line.catalog_url', 'cell_line.cell_cycle.text', 'cell_line.cell_cycle.ontology', 'cell_line.cell_cycle.ontology_label', 'cell_line.cell_line_type', 'cell_line.model_organ.text', 'cell_line.model_organ.ontology', 'cell_line.model_organ.ontology_label', 'cell_line.cell_morphology.cell_morphology', 'cell_line.cell_morphology.cell_size', 'cell_line.cell_morphology.cell_size_unit.text', 'cell_line.cell_morphology.cell_size_unit.ontology', 'cell_line.cell_morphology.cell_size_unit.ontology_label', 'cell_line.cell_morphology.percent_cell_viability', 'cell_line.cell_morphology.cell_viability_method', 'cell_line.cell_morphology.cell_viability_result', 'cell_line.cell_morphology.percent_necrosis', 'cell_line.growth_conditions.passage_number', 'cell_line.growth_conditions.growth_medium', 'cell_line.growth_conditions.culture_environment', 'cell_line.growth_conditions.mycoplasma_testing_method', 'cell_line.growth_conditions.mycoplasma_testing_results', 'cell_line.growth_conditions.drug_treatment', 'cell_line.growth_conditions.feeder_layer_type', 'cell_line.confluency', 'cell_line.cell_type.text', 'cell_line.cell_type.ontology', 'cell_line.cell_type.ontology_label', 'cell_line.karyotype', 'cell_line.tissue.text', 'cell_line.tissue.ontology', 'cell_line.tissue.ontology_label', 'cell_line.date_established', 'cell_line.disease.text', 'cell_line.disease.ontology', 'cell_line.disease.ontology_label', 'cell_line.genus_species.text', 'cell_line.genus_species.ontology', 'cell_line.genus_species.ontology_label', 'cell_line.publication.authors', 'cell_line.publication.publication_title', 'cell_line.publication.doi', 'cell_line.publication.pmid', 'cell_line.publication.publication_url']\n"
     ]
    }
   ],
   "source": [
    "# get latest attributes from schema\n",
    "\n",
    "from ingest.template.schema_template import SchemaTemplate\n",
    "import yaml\n",
    "\n",
    "template = SchemaTemplate()\n",
    "attribute_yaml = template.yaml_dump()\n",
    "\n",
    "attribute_list = yaml.load(attribute_yaml).get('tabs')\n",
    "full_set = set()\n",
    "full_list = []\n",
    "for tab in attribute_list:\n",
    "    for x in list(tab.values()):\n",
    "        for attribute in x.get('columns'):\n",
    "            full_set.add(attribute)\n",
    "            full_list.append(attribute)\n",
    "print('{} out of {} attributes are unique'.format(len(full_set), len(full_list)))\n",
    "print(full_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "553dfbe7ffcf434498c5b10160172f29",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=3), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "3 of 3 bundles processed\n"
     ]
    }
   ],
   "source": [
    "import hca.dss\n",
    "from hca.dss import DSSClient\n",
    "import sys\n",
    "import json\n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "\n",
    "project_q = {\n",
    "    \"query\": {\n",
    "        \"bool\": {\n",
    "            \"must\": [\n",
    "                {\n",
    "                    \"terms\": {\n",
    "                        \"files.project_json.provenance.document_id\": [\n",
    "                            \"PROJECT_UUID\"\n",
    "                        ]\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "q = json.loads(str(project_q).replace('PROJECT_UUID', query_uuid).replace(\"'\", '\"'))\n",
    "\n",
    "\n",
    "dss_client = DSSClient(swagger_url= \"https://dss.data.humancellatlas.org/v1/swagger.json\")\n",
    "bundle_generator = dss_client.post_search.iterate(replica=\"aws\", es_query=q, output_format=\"raw\")\n",
    "total_hits = dss_client.post_search(replica=\"aws\", es_query=q, output_format=\"raw\").get('total_hits')\n",
    "\n",
    "bundle_attributes = []\n",
    "\n",
    "for bundle in tqdm(bundle_generator, total=total_hits, unit='bundle'):\n",
    "    bundle_meta = bundle.get('metadata').get('files')\n",
    "    top_level = list(bundle_meta.keys())\n",
    "   \n",
    "    if len(top_level) == 0:\n",
    "        print('WARN: Bundle had empty metadata') # check for erronious bundles\n",
    "        sys.exit()\n",
    "    \n",
    "    detected_attributes = []\n",
    "    \n",
    "    # certain branches can be ignored when exploring the tree.\n",
    "    ignore_top_level = ['links_json', 'analysis_process']\n",
    "    ignore_mid_level = ['schema_type', 'provenance', 'describedBy']\n",
    "    \n",
    "    \n",
    "        \n",
    "    for x in top_level:\n",
    "        if x in ignore_top_level: # skip these fields\n",
    "            continue\n",
    "        for meta_doc in bundle_meta.get(x):\n",
    "            top_level = x[:-5] # strip '_json'\n",
    "            for mid_level, value in meta_doc.items():\n",
    "                if mid_level in ignore_mid_level:\n",
    "                    continue\n",
    "                if type(value) != list:\n",
    "                    value_list = [value]\n",
    "                else:\n",
    "                    value_list = value\n",
    "                for item in value_list:     \n",
    "                    if (type(item) == str) or (type(item) == int) or (type(item) == bool):\n",
    "                        attribute = '.'.join([top_level, mid_level])\n",
    "                        detected_attributes.append(attribute)\n",
    "                        continue\n",
    "                    for low_level in item.items():\n",
    "                        if low_level in ignore_mid_level:\n",
    "                            continue\n",
    "                        attribute = '.'.join([top_level, mid_level, low_level[0]])\n",
    "                        detected_attributes.append(attribute)\n",
    "\n",
    "\n",
    "    bundle_attributes.append(detected_attributes)\n",
    "print('{} of {} bundles processed'.format(len(bundle_attributes), total_hits))\n",
    "project_attribute_list = set([item for sublist in bundle_attributes for item in sublist])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 out of 74 attributes do not meet latest schema requirements.\n",
      "This project's metadata is OUT OF DATE\n",
      "See replacement suggestions here:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Project attributes</th>\n",
       "      <th>Closest match in latest schema</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sequencing_protocol.sequencing_approach.text</td>\n",
       "      <td>sequencing_protocol.method.text</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dissociation_protocol.dissociation_method.text</td>\n",
       "      <td>dissociation_protocol.method.text</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cell_suspension.total_estimated_cells</td>\n",
       "      <td>cell_suspension.estimated_cell_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>library_preparation_protocol.library_construct...</td>\n",
       "      <td>library_preparation_protocol.library_construct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>library_preparation_protocol.library_construct...</td>\n",
       "      <td>library_preparation_protocol.library_construct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sequencing_protocol.sequencing_approach.ontology</td>\n",
       "      <td>sequencing_protocol.method.ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sequencing_protocol.sequencing_approach.ontolo...</td>\n",
       "      <td>sequencing_protocol.method.ontology_label</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>dissociation_protocol.dissociation_method.onto...</td>\n",
       "      <td>dissociation_protocol.method.ontology_label</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>library_preparation_protocol.library_construct...</td>\n",
       "      <td>library_preparation_protocol.library_construct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dissociation_protocol.dissociation_method.onto...</td>\n",
       "      <td>dissociation_protocol.method.ontology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Project attributes  \\\n",
       "0       sequencing_protocol.sequencing_approach.text   \n",
       "1     dissociation_protocol.dissociation_method.text   \n",
       "2              cell_suspension.total_estimated_cells   \n",
       "3  library_preparation_protocol.library_construct...   \n",
       "4  library_preparation_protocol.library_construct...   \n",
       "5   sequencing_protocol.sequencing_approach.ontology   \n",
       "6  sequencing_protocol.sequencing_approach.ontolo...   \n",
       "7  dissociation_protocol.dissociation_method.onto...   \n",
       "8  library_preparation_protocol.library_construct...   \n",
       "9  dissociation_protocol.dissociation_method.onto...   \n",
       "\n",
       "                      Closest match in latest schema  \n",
       "0                    sequencing_protocol.method.text  \n",
       "1                  dissociation_protocol.method.text  \n",
       "2               cell_suspension.estimated_cell_count  \n",
       "3  library_preparation_protocol.library_construct...  \n",
       "4  library_preparation_protocol.library_construct...  \n",
       "5                sequencing_protocol.method.ontology  \n",
       "6          sequencing_protocol.method.ontology_label  \n",
       "7        dissociation_protocol.method.ontology_label  \n",
       "8  library_preparation_protocol.library_construct...  \n",
       "9              dissociation_protocol.method.ontology  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import difflib\n",
    "import pandas as pd\n",
    "\n",
    "out_dated_attributes = list(set(project_attribute_list) - set(full_list))\n",
    "shared_attributes = list(set(project_attribute_list).intersection(full_list))\n",
    "unused = list(set(full_list) - set(project_attribute_list)) # unused attributes in latest schema\n",
    "\n",
    "# print('This project did NOT use {} attributes from the latest schemas.'.format(len(unused)))\n",
    "# print('{} attributes in the project were also found in the latest schemas.'.format(len(shared_attributes)))\n",
    "# print('{} attributes were present in the project that are not part of the latest schemas.'.format(len(out_dated_attributes)))\n",
    "\n",
    "print('{} out of {} attributes do not meet latest schema requirements.'.format(len(out_dated_attributes), len(shared_attributes)))\n",
    "if len(out_dated_attributes) > 0:\n",
    "    print(\"This project's metadata is OUT OF DATE\")\n",
    "else:\n",
    "    print(\"This project's metadata is UP TO DATE\")\n",
    "\n",
    "# print(out_dated_attributes)\n",
    "matches = {}\n",
    "for ood_attribute in out_dated_attributes:\n",
    "    closest = difflib.get_close_matches(ood_attribute, full_list, n=1)[0]\n",
    "\n",
    "#     print('Project has: {}'.format(ood_attribute))\n",
    "#     print('Closest match in latest schema is: {}'.format(closest))\n",
    "\n",
    "    matches[ood_attribute] = closest\n",
    "    \n",
    "df = pd.DataFrame.from_dict(matches, orient='index').reset_index().rename(index=str, columns={\"index\": \"Project attributes\", 0: \"Closest match in latest schema\"})\n",
    "print('See replacement suggestions here:')\n",
    "outfile = 'runs/' + query_uuid + '_diff.csv'\n",
    "df.to_csv(outfile) # unhash to saveout a result\n",
    "df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO\n",
    "- Needs to catch 4th nested attributes e.g. donor_organism.mouse_specific.strain.text\n",
    "- analysis_process attributes are leaking through still"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results from cBeta\n",
    "\n",
    "The following terms have been replaced\n",
    "\n",
    "| Out of date attribute                                                     | Suggested replacement                                                   | \n",
    "|---------------------------------------------------------------------------|-------------------------------------------------------------------------| \n",
    "| dissociation_protocol.dissociation_method.ontology                        | dissociation_protocol.method.ontology                                   | \n",
    "| dissociation_protocol.dissociation_method.ontology_label                  | dissociation_protocol.method.ontology_label                             | \n",
    "| library_preparation_protocol.library_construction_approach.text           | library_preparation_protocol.library_construction_method.text           | \n",
    "| sequencing_protocol.sequencing_approach.text                              | sequencing_protocol.method.text                                         | \n",
    "| dissociation_protocol.dissociation_method.text                            | dissociation_protocol.method.text                                       | \n",
    "| sequencing_protocol.sequencing_approach.ontology                          | sequencing_protocol.method.ontology                                     | \n",
    "| specimen_from_organism.organ_part.text                                    | specimen_from_organism.organ_parts.text                                 | \n",
    "| specimen_from_organism.organ_part.ontology_label                          | specimen_from_organism.organ_parts.ontology_label                       | \n",
    "| library_preparation_protocol.library_construction_approach.ontology_label | library_preparation_protocol.library_construction_method.ontology_label | \n",
    "| specimen_from_organism.organ_part.ontology                                | specimen_from_organism.organ_parts.ontology                             | \n",
    "| sequencing_protocol.sequencing_approach.ontology_label                    | sequencing_protocol.method.ontology_label                               | \n",
    "| library_preparation_protocol.library_construction_approach.ontology       | library_preparation_protocol.library_construction_method.ontology       | \n",
    "| sequence_file.insdc_run                                                   | sequence_file.insdc_run_accessions                                      | \n",
    "| cell_suspension.total_estimated_cells                                     | cell_suspension.estimated_cell_count                                    | \n",
    "| project.insdc_project                                                     | project.insdc_project_accessions                                        | \n",
    "| project.geo_series                                                        | project.geo_series_accessions                                           | \n",
    "| project.funders.funder_name                                               | ?                                                | \n",
    "| process.insdc_experiment.insdc_experiment                                 | process.insdc_experiment.insdc_experiment_accession                     | \n",
    "| cell_suspension.plate_based_sequencing.cell_quality                       | cell_suspension.plate_based_sequencing.well_quality                     | \n",
    "| collection_protocol.collection_method.ontology                            | collection_protocol.method.ontology                                     | \n",
    "| project.insdc_study                                                       | project.insdc_study_accessions                                          | \n",
    "| collection_protocol.collection_method.text                                | collection_protocol.method.text                                         | \n",
    "| collection_protocol.collection_method.ontology_label                      | collection_protocol.method.ontology_label                               | \n",
    "| cell_suspension.plate_based_sequencing.plate_id                           | cell_suspension.plate_based_sequencing.plate_label                      | \n",
    "| project.array_express_investigation                                       | project.array_express_accessions                                        | \n",
    "| cell_suspension.plate_based_sequencing.well_id                            | cell_suspension.plate_based_sequencing.well_quality                     | \n",
    "| donor_organism.familial_relationship.child                                | donor_organism.familial_relationships.child                             | \n",
    "| donor_organism.familial_relationship.parent                               | donor_organism.familial_relationships.parent                            | \n",
    "| dissociation_protocol.protocol_reagents.catalog_number                    | ipsc_induction_protocol.protocol_reagents.catalog_number                | \n",
    "| dissociation_protocol.protocol_reagents.manufacturer                      | ipsc_induction_protocol.protocol_reagents.manufacturer                  | \n",
    "| dissociation_protocol.protocol_reagents.expiry_date                       | ipsc_induction_protocol.protocol_reagents.expiry_date                   | \n",
    "| dissociation_protocol.protocol_reagents.lot_number                        | ipsc_induction_protocol.protocol_reagents.lot_number                    | \n",
    "| dissociation_protocol.protocol_reagents.retail_name                       | ipsc_induction_protocol.protocol_reagents.retail_name                   | \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
